// @inkshot/engine/dist/plugins/tilemap/TiledLoader.js
// ---------------------------------------------------------------------------
// TiledLoader
// ---------------------------------------------------------------------------
/**
 * Convert a Tiled JSON map (`.tmj`) into `TilemapData` and `ActorDef[]`.
 *
 * ### Usage
 * ```ts
 * import { loadTiledMap } from 'inkshot-engine';
 *
 * const raw = await fetch('/maps/level1.tmj').then(r => r.json());
 * const { tilemapData, actors } = loadTiledMap(raw, {
 *   tilesetTextureKeys: { 'world': 'tileset_world' },
 *   actorFactories: {
 *     'Player': (obj, props) => ({ id: 'player', type: 'player', ...props }),
 *     'Enemy':  (obj, props) => ({ id: `enemy_${obj.id}`, type: 'enemy', ...props }),
 *   },
 * });
 *
 * await core.events.emit('tilemap/load', { mapData: tilemapData });
 * for (const actor of actors) {
 *   core.events.emitSync('actor/define', actor);
 * }
 * ```
 *
 * @param map  Parsed Tiled JSON object (result of `JSON.parse(tmjString)`).
 * @param opts Conversion options.
 */
export function loadTiledMap(map, opts) {
    const collisionPropMap = opts.collisionPropertyMap ?? { collision: 'solid' };
    // ── 1. Convert tilesets ──────────────────────────────────────────────────
    const tilesets = [];
    for (const ref of map.tilesets) {
        const name = ref.name ?? '';
        const textureKey = ref.textureKey ?? opts.tilesetTextureKeys[name] ?? '';
        tilesets.push({
            firstgid: ref.firstgid,
            name,
            textureKey,
            tileWidth: ref.tilewidth ?? map.tilewidth,
            tileHeight: ref.tileheight ?? map.tileheight,
            columns: ref.columns,
            spacing: ref.spacing,
            margin: ref.margin,
        });
    }
    // ── 2. Build per-GID tile metadata (collision shapes, animation) ─────────
    //   (We walk all inline tileset objects from the raw map if present.)
    const gidToShape = new Map();
    const animatedTiles = {};
    for (const ref of map.tilesets) {
        const firstgid = ref.firstgid;
        // Inline tile data is stored in `ref` when the tileset is embedded.
        const tiles = ref.tiles;
        if (!tiles)
            continue;
        for (const tile of tiles) {
            const gid = firstgid + tile.id;
            // Collision shape from custom properties.
            if (tile.properties) {
                for (const prop of tile.properties) {
                    if (prop.name in collisionPropMap && prop.value) {
                        gidToShape.set(gid, collisionPropMap[prop.name]);
                        break;
                    }
                }
            }
            // Animation frames.
            if (tile.animation && tile.animation.length > 0) {
                const frames = tile.animation.map((f) => ({
                    tileId: firstgid + f.tileid,
                    duration: f.duration,
                }));
                animatedTiles[gid] = { frames };
            }
        }
    }
    // ── 3. Convert tile layers ────────────────────────────────────────────────
    const tilemapLayers = [];
    const actors = [];
    const autotileGroups = [];
    _flattenLayers(map.layers, tilemapLayers, actors, opts, gidToShape);
    // ── 4. Assemble TilemapData ───────────────────────────────────────────────
    const tilemapData = {
        tileWidth: map.tilewidth,
        tileHeight: map.tileheight,
        mapWidth: map.width,
        mapHeight: map.height,
        tilesets,
        layers: tilemapLayers,
        ...(Object.keys(animatedTiles).length > 0 ? { animatedTiles } : {}),
        ...(autotileGroups.length > 0 ? { autotileGroups } : {}),
        ...(opts.chunkSize !== undefined ? { chunkSize: opts.chunkSize } : {}),
    };
    return { tilemapData, actors };
}
// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------
function _flattenLayers(layers, outTileLayers, outActors, opts, gidToShape) {
    for (const layer of layers) {
        if (layer.type === 'group') {
            _flattenLayers(layer.layers, outTileLayers, outActors, opts, gidToShape);
            continue;
        }
        if (layer.type === 'tilelayer') {
            outTileLayers.push(_convertTileLayer(layer, opts, gidToShape));
            continue;
        }
        if (layer.type === 'objectgroup' && opts.actorFactories) {
            for (const obj of layer.objects) {
                const factory = opts.actorFactories[obj.type];
                if (factory) {
                    const props = _parseProperties(obj.properties ?? []);
                    outActors.push(factory(obj, props));
                }
            }
        }
    }
}
function _convertTileLayer(layer, opts, gidToShape) {
    const isCollisionLayer = opts.autoCollisionLayers === true &&
        layer.name.toLowerCase().startsWith('collision');
    // Build the tileShapes map for this layer.
    let tileShapes;
    if (isCollisionLayer || _layerHasCollisionTiles(layer, gidToShape)) {
        tileShapes = {};
        for (const gid of layer.data) {
            if (gid !== 0 && gidToShape.has(gid) && !(gid in tileShapes)) {
                tileShapes[gid] = gidToShape.get(gid);
            }
        }
        if (isCollisionLayer) {
            // Mark every non-empty tile as solid when it doesn't have an explicit shape.
            for (const gid of layer.data) {
                if (gid !== 0 && !(gid in tileShapes)) {
                    tileShapes[gid] = 'solid';
                }
            }
        }
        if (Object.keys(tileShapes).length === 0)
            tileShapes = undefined;
    }
    const result = {
        name: layer.name,
        data: layer.data,
        visible: layer.visible ?? true,
        opacity: layer.opacity ?? 1,
    };
    if (tileShapes && Object.keys(tileShapes).length > 0) {
        result.collider = true;
        result.tileShapes = tileShapes;
    }
    return result;
}
function _layerHasCollisionTiles(layer, gidToShape) {
    for (const gid of layer.data) {
        if (gid !== 0 && gidToShape.has(gid))
            return true;
    }
    return false;
}
function _parseProperties(props) {
    const result = {};
    for (const p of props) {
        result[p.name] = p.value;
    }
    return result;
}
