// @inkshot/engine/dist/plugins/save/IndexedDBSaveAdapter.js
// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------
const SLOT_STORE = 'slots';
const GLOBAL_STORE = 'global';
const GLOBAL_KEY = '__global__';
// ---------------------------------------------------------------------------
// Real IndexedDB implementation
// ---------------------------------------------------------------------------
/**
 * Opens a real browser IndexedDB database and exposes the two required object
 * stores (`slots` and `global`) through the {@link IDBStoreLike} interface.
 *
 * @internal
 */
class RealIDBStore {
    _name;
    _version;
    _db = null;
    constructor(_name, _version) {
        this._name = _name;
        this._version = _version;
    }
    /** Open the database and create object stores if needed. */
    open() {
        return new Promise((resolve, reject) => {
            const req = globalThis.indexedDB.open(this._name, this._version);
            req.onupgradeneeded = (event) => {
                const db = event.target.result;
                if (!db.objectStoreNames.contains(SLOT_STORE)) {
                    db.createObjectStore(SLOT_STORE);
                }
                if (!db.objectStoreNames.contains(GLOBAL_STORE)) {
                    db.createObjectStore(GLOBAL_STORE);
                }
            };
            req.onsuccess = () => {
                this._db = req.result;
                resolve();
            };
            req.onerror = () => reject(req.error);
        });
    }
    /** Close the underlying database connection. */
    close() {
        this._db?.close();
        this._db = null;
    }
    get(storeName, key) {
        return new Promise((resolve, reject) => {
            const tx = this._db.transaction(storeName, 'readonly');
            const req = tx.objectStore(storeName).get(key);
            req.onsuccess = () => resolve(req.result);
            req.onerror = () => reject(req.error);
        });
    }
    put(storeName, key, value) {
        return new Promise((resolve, reject) => {
            const tx = this._db.transaction(storeName, 'readwrite');
            const req = tx.objectStore(storeName).put(value, key);
            req.onsuccess = () => resolve();
            req.onerror = () => reject(req.error);
        });
    }
    delete(storeName, key) {
        return new Promise((resolve, reject) => {
            const tx = this._db.transaction(storeName, 'readwrite');
            const req = tx.objectStore(storeName).delete(key);
            req.onsuccess = () => resolve();
            req.onerror = () => reject(req.error);
        });
    }
}
// ---------------------------------------------------------------------------
// IndexedDBSaveAdapter
// ---------------------------------------------------------------------------
/**
 * Browser-side persistence adapter for the {@link SaveManager} that uses
 * **IndexedDB** as the backing store.
 *
 * IndexedDB serialises values via the _Structured Clone_ algorithm, so it
 * supports binary data (`ArrayBuffer`, `Blob`, `TypedArray`) and complex
 * objects in addition to plain JSON-compatible values, making it better suited
 * than `localStorage` for large or binary save data.
 *
 * `IndexedDBSaveAdapter` mirrors the event contract of
 * {@link LocalStorageSaveAdapter}:
 *
 * | Phase  | Event             | What this adapter does                                   |
 * |--------|-------------------|----------------------------------------------------------|
 * | before | `save/slot:load`  | Reads entry from IDB → sets `output.raw`                 |
 * | after  | `save/slot:save`  | Writes `output.data` to IDB                              |
 * | after  | `save/slot:delete`| Removes the slot entry from IDB                         |
 * | before | `save/global:load`| Reads entry from IDB → sets `output.raw`                 |
 * | after  | `save/global:save`| Writes `output.data` to IDB                              |
 *
 * ### Usage
 *
 * ```ts
 * import { createEngine, SaveManager, IndexedDBSaveAdapter } from 'inkshot-engine';
 *
 * const { core } = await createEngine({
 *   plugins: [
 *     new SaveManager(),
 *     new IndexedDBSaveAdapter({ dbName: 'my-game' }),
 *   ],
 * });
 *
 * // Persists to IndexedDB
 * await core.events.emit('save/slot:set',  { id: 'slot-1', patch: { level: 3 } });
 * await core.events.emit('save/slot:save', { id: 'slot-1' });
 *
 * // Restores from IndexedDB
 * await core.events.emit('save/slot:load', { id: 'slot-1' });
 * ```
 */
export class IndexedDBSaveAdapter {
    namespace = 'indexedDbSave';
    editorMeta = {
        displayName: 'IndexedDB Save Adapter',
        icon: 'save',
        description: 'Persists save data to the browser IndexedDB for larger storage capacity.',
        events: [
            'save/slot:save', 'save/slot:load', 'save/slot:delete',
            'save/global:save', 'save/global:load',
        ],
    };
    _dbName;
    _dbVersion;
    _injectedStore;
    _store = null;
    _realStore = null;
    constructor(options = {}) {
        this._dbName = options.dbName ?? 'inkshot-saves';
        this._dbVersion = options.dbVersion ?? 1;
        this._injectedStore = options.idbStore;
    }
    // ---------------------------------------------------------------------------
    // EnginePlugin lifecycle
    // ---------------------------------------------------------------------------
    async init(core) {
        if (this._injectedStore) {
            this._store = this._injectedStore;
        }
        else {
            const real = new RealIDBStore(this._dbName, this._dbVersion);
            await real.open();
            this._realStore = real;
            this._store = real;
        }
        const store = this._store;
        const { events } = core;
        // ── Slot: save (after) ───────────────────────────────────────────────────
        events.on(this.namespace, 'save/slot:save', async (_params, output) => {
            if (output.data !== undefined) {
                try {
                    await store.put(SLOT_STORE, output.data.meta.id, output.data);
                    output.saved = true;
                }
                catch (err) {
                    console.warn('[IndexedDBSaveAdapter] Failed to persist slot:', err);
                }
            }
        }, { phase: 'after' });
        // ── Slot: load (before) ──────────────────────────────────────────────────
        events.on(this.namespace, 'save/slot:load', async (params, output) => {
            try {
                const data = await store.get(SLOT_STORE, params.id);
                if (data !== undefined) {
                    output.raw = data;
                }
            }
            catch (err) {
                console.warn(`[IndexedDBSaveAdapter] Failed to read slot "${params.id}":`, err);
            }
        }, { phase: 'before' });
        // ── Slot: delete (after) ─────────────────────────────────────────────────
        events.on(this.namespace, 'save/slot:delete', async (params) => {
            try {
                await store.delete(SLOT_STORE, params.id);
            }
            catch (err) {
                console.warn(`[IndexedDBSaveAdapter] Failed to delete slot "${params.id}":`, err);
            }
        }, { phase: 'after' });
        // ── Global: save (after) ─────────────────────────────────────────────────
        events.on(this.namespace, 'save/global:save', async (_params, output) => {
            if (output.data !== undefined) {
                try {
                    await store.put(GLOBAL_STORE, GLOBAL_KEY, output.data);
                    output.saved = true;
                }
                catch (err) {
                    console.warn('[IndexedDBSaveAdapter] Failed to persist global data:', err);
                }
            }
        }, { phase: 'after' });
        // ── Global: load (before) ────────────────────────────────────────────────
        events.on(this.namespace, 'save/global:load', async (_params, output) => {
            try {
                const data = await store.get(GLOBAL_STORE, GLOBAL_KEY);
                if (data !== undefined) {
                    output.raw = data;
                }
            }
            catch (err) {
                console.warn('[IndexedDBSaveAdapter] Failed to read global data:', err);
            }
        }, { phase: 'before' });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
        this._realStore?.close();
        this._realStore = null;
        this._store = null;
    }
}
