// @inkshot/engine/dist/plugins/save/LocalStorageSaveAdapter.js
// ---------------------------------------------------------------------------
// LocalStorageSaveAdapter
// ---------------------------------------------------------------------------
/**
 * Browser-side persistence adapter for the {@link SaveManager}.
 *
 * `LocalStorageSaveAdapter` is the **storage layer** of the save system.  It
 * listens to the same events as `SaveManager` but in the complementary phases,
 * serialising and deserialising data to/from `localStorage`:
 *
 * | Phase  | Event             | What this adapter does                           |
 * |--------|-------------------|--------------------------------------------------|
 * | before | `save/slot:load`  | Reads JSON from storage → sets `output.raw`      |
 * | after  | `save/slot:save`  | Writes `output.data` as JSON to storage          |
 * | after  | `save/slot:delete`| Removes the slot key from storage               |
 * | before | `save/global:load`| Reads JSON from storage → sets `output.raw`      |
 * | after  | `save/global:save`| Writes `output.data` as JSON to storage          |
 *
 * ### Usage
 *
 * ```ts
 * import { createEngine, SaveManager, LocalStorageSaveAdapter } from 'inkshot-engine';
 *
 * const { core } = await createEngine({
 *   plugins: [
 *     new SaveManager(),
 *     new LocalStorageSaveAdapter({ keyPrefix: 'mygame:' }),
 *   ],
 * });
 *
 * // Persists to localStorage under the key 'mygame:slot:slot-1'
 * await core.events.emit('save/slot:set',  { id: 'slot-1', patch: { level: 3 } });
 * await core.events.emit('save/slot:save', { id: 'slot-1' });
 *
 * // Restores from localStorage automatically via the before-phase hook
 * await core.events.emit('save/slot:load', { id: 'slot-1' });
 * ```
 */
export class LocalStorageSaveAdapter {
    namespace = 'localStorageSave';
    editorMeta = {
        displayName: 'LocalStorage Save Adapter',
        icon: 'save',
        description: 'Persists save data to the browser localStorage.',
        events: [
            'save/slot:save', 'save/slot:load', 'save/slot:delete',
            'save/global:save', 'save/global:load',
        ],
    };
    _prefix;
    _storage;
    constructor(options = {}) {
        this._prefix = options.keyPrefix ?? 'inkshot:';
        this._storage = options.storage ?? globalThis.localStorage;
    }
    // ---------------------------------------------------------------------------
    // EnginePlugin lifecycle
    // ---------------------------------------------------------------------------
    init(core) {
        const { events } = core;
        // ── Slot: save (after) ───────────────────────────────────────────────────
        events.on(this.namespace, 'save/slot:save', (_params, output) => {
            if (output.data !== undefined) {
                try {
                    this._storage.setItem(this._slotKey(output.data.meta.id), JSON.stringify(output.data));
                    output.saved = true;
                }
                catch (err) {
                    console.warn('[LocalStorageSaveAdapter] Failed to persist slot:', err);
                }
            }
        }, { phase: 'after' });
        // ── Slot: load (before) ──────────────────────────────────────────────────
        events.on(this.namespace, 'save/slot:load', (params, output) => {
            const raw = this._storage.getItem(this._slotKey(params.id));
            if (raw !== null) {
                try {
                    output.raw = JSON.parse(raw);
                }
                catch (err) {
                    console.warn(`[LocalStorageSaveAdapter] Corrupt slot data for "${params.id}":`, err);
                }
            }
        }, { phase: 'before' });
        // ── Slot: delete (after) ─────────────────────────────────────────────────
        events.on(this.namespace, 'save/slot:delete', (params) => {
            this._storage.removeItem(this._slotKey(params.id));
        }, { phase: 'after' });
        // ── Global: save (after) ─────────────────────────────────────────────────
        events.on(this.namespace, 'save/global:save', (_params, output) => {
            if (output.data !== undefined) {
                try {
                    this._storage.setItem(this._globalKey(), JSON.stringify(output.data));
                    output.saved = true;
                }
                catch (err) {
                    console.warn('[LocalStorageSaveAdapter] Failed to persist global data:', err);
                }
            }
        }, { phase: 'after' });
        // ── Global: load (before) ────────────────────────────────────────────────
        events.on(this.namespace, 'save/global:load', (_params, output) => {
            const raw = this._storage.getItem(this._globalKey());
            if (raw !== null) {
                try {
                    output.raw = JSON.parse(raw);
                }
                catch (err) {
                    console.warn('[LocalStorageSaveAdapter] Corrupt global save data:', err);
                }
            }
        }, { phase: 'before' });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
    }
    // ---------------------------------------------------------------------------
    // Private helpers
    // ---------------------------------------------------------------------------
    _slotKey(id) {
        return `${this._prefix}slot:${id}`;
    }
    _globalKey() {
        return `${this._prefix}global`;
    }
}
