// @inkshot/engine/dist/plugins/save/SaveMigrationPlugin.js
// ---------------------------------------------------------------------------
// SaveMigrationPlugin
// ---------------------------------------------------------------------------
/**
 * Plugin that provides **save-data version migration** for the
 * {@link SaveManager}.
 *
 * ### How it works
 *
 * **On save** (`save/slot:save` / `save/global:save` — `after` phase,
 * priority `1`):
 * The plugin stamps the `currentVersion` onto `output.data` _before_ the
 * environment adapter (e.g. {@link LocalStorageSaveAdapter} or
 * {@link IndexedDBSaveAdapter}) writes it to persistent storage.  This
 * ensures every persisted entry carries an explicit schema version.
 *
 * **On load** (`save/slot:load` / `save/global:load` — `before` phase,
 * priority `-1`):
 * After the environment adapter has placed the raw entry into `output.raw`,
 * the plugin inspects its stored version and applies each
 * {@link SaveMigrationStep} in sequence (from `storedVersion` up to
 * `currentVersion − 1`) before `SaveManager` writes the data into memory.
 * Save data with no version field is treated as version `0`.
 *
 * ### Priority ordering
 *
 * | Phase  | Event              | Listener                  | Priority |
 * |--------|--------------------|---------------------------|----------|
 * | after  | `save/slot:save`   | SaveMigrationPlugin stamp | `1`      |
 * | after  | `save/slot:save`   | env adapter persist       | `0`      |
 * | before | `save/slot:load`   | env adapter read          | `0`      |
 * | before | `save/slot:load`   | SaveMigrationPlugin apply | `-1`     |
 *
 * The same ordering applies to `save/global:save` / `save/global:load`.
 *
 * ### Usage
 *
 * ```ts
 * import {
 *   createEngine, SaveManager, LocalStorageSaveAdapter,
 *   SaveMigrationPlugin,
 * } from 'inkshot-engine';
 *
 * const { core } = await createEngine({
 *   plugins: [
 *     new SaveManager(),
 *     new LocalStorageSaveAdapter(),
 *     new SaveMigrationPlugin({
 *       currentVersion: 2,
 *       migrations: [
 *         {
 *           fromVersion: 0,
 *           slot: (data) => {
 *             // v0 → v1: rename `gold` to `coins`
 *             if (typeof data['gold'] === 'number') {
 *               data['coins'] = data['gold'];
 *               delete data['gold'];
 *             }
 *             return data;
 *           },
 *         },
 *         {
 *           fromVersion: 1,
 *           slot: (data) => {
 *             // v1 → v2: initialise new `level` field
 *             data['level'] ??= 1;
 *             return data;
 *           },
 *         },
 *       ],
 *     }),
 *   ],
 * });
 * ```
 */
export class SaveMigrationPlugin {
    namespace = 'saveMigration';
    editorMeta = {
        displayName: 'Save Migration Plugin',
        icon: 'save',
        description: 'Applies versioned migration steps to save data during load.',
        events: [],
    };
    _currentVersion;
    _migrations;
    constructor(options) {
        this._currentVersion = options.currentVersion;
        this._migrations = new Map([...options.migrations].sort((a, b) => a.fromVersion - b.fromVersion).map((s) => [s.fromVersion, s]));
    }
    // ---------------------------------------------------------------------------
    // EnginePlugin lifecycle
    // ---------------------------------------------------------------------------
    init(core) {
        const { events } = core;
        // ── Stamp version on save (after phase, priority 1 — runs before env adapter) ──
        events.on(this.namespace, 'save/slot:save', (_params, output) => {
            if (output.data?.meta !== undefined) {
                output.data.meta.version = this._currentVersion;
            }
        }, { phase: 'after', priority: 1 });
        events.on(this.namespace, 'save/global:save', (_params, output) => {
            if (output.data !== undefined) {
                output.data.version = this._currentVersion;
            }
        }, { phase: 'after', priority: 1 });
        // ── Apply migrations on load (before phase, priority -1 — runs after env adapter) ──
        events.on(this.namespace, 'save/slot:load', (_params, output) => {
            if (output.raw !== undefined) {
                output.raw = this._migrateSlot(output.raw);
            }
        }, { phase: 'before', priority: -1 });
        events.on(this.namespace, 'save/global:load', (_params, output) => {
            if (output.raw !== undefined) {
                output.raw = this._migrateGlobal(output.raw);
            }
        }, { phase: 'before', priority: -1 });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
    }
    // ---------------------------------------------------------------------------
    // Private helpers
    // ---------------------------------------------------------------------------
    /**
     * Apply all pending slot migration steps from the data's stored version up
     * to `currentVersion`, returning a new (deep-cloned) `SlotData` object.
     */
    _migrateSlot(data) {
        let version = data.meta.version ?? 0;
        if (version >= this._currentVersion)
            return data;
        const cloned = structuredClone(data);
        while (version < this._currentVersion) {
            const step = this._migrations.get(version);
            if (step?.slot) {
                const result = step.slot(cloned.data);
                if (result !== undefined) {
                    cloned.data = result;
                }
            }
            version++;
        }
        cloned.meta.version = this._currentVersion;
        return cloned;
    }
    /**
     * Apply all pending global migration steps from the data's stored version up
     * to `currentVersion`, returning a new (deep-cloned) `GlobalSaveData` object.
     */
    _migrateGlobal(data) {
        let version = data.version ?? 0;
        if (version >= this._currentVersion)
            return data;
        const cloned = structuredClone(data);
        while (version < this._currentVersion) {
            const step = this._migrations.get(version);
            if (step?.global) {
                const result = step.global(cloned.data);
                if (result !== undefined) {
                    cloned.data = result;
                }
            }
            version++;
        }
        cloned.version = this._currentVersion;
        return cloned;
    }
}
