// @inkshot/engine/dist/plugins/rpg/BattleSystem.js
// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------
// BattleSystem
// ---------------------------------------------------------------------------
/**
 * Plugin that manages **turn-based battle sessions**.
 *
 * Multiple battles can run concurrently (each identified by a `battleId`).
 * The phase flow is:
 * ```
 * acting → resolving → (next turn: acting) → end
 * ```
 *
 * ### EventBus API
 *
 * | Event                    | Params / Output                                |
 * |--------------------------|------------------------------------------------|
 * | `battle/start`           | `BattleStartParams → BattleStartOutput`        |
 * | `battle/action`          | `BattleActionParams`                           |
 * | `battle/resolve`         | `BattleResolveParams → BattleResolveOutput`    |
 * | `battle/end`             | `BattleEndParams`                              |
 * | `battle/state:get`       | `BattleStateGetParams → BattleStateGetOutput`  |
 * | `battle/started`         | `BattleStartedParams` (notification)           |
 * | `battle/turn:start`      | `BattleTurnStartParams` (notification)         |
 * | `battle/action:resolved` | `BattleActionResolvedParams` (notification)    |
 * | `battle/ended`           | `BattleEndedParams` (notification)             |
 * | `battle/combatant:defeated` | `BattleCombatantDefeatedParams` (notification)|
 */
export class BattleSystem {
    namespace = 'battle';
    dependencies = ['stats', 'inventory'];
    editorMeta = {
        displayName: 'Battle System',
        icon: 'battle',
        description: 'Turn-based battle system with action resolution and combatant defeat tracking.',
        events: [
            'battle/start', 'battle/end', 'battle/action', 'battle/resolve', 'battle/state:get',
        ],
    };
    _battles = new Map();
    _battleCounter = 0;
    _generateBattleId() {
        return `battle_${++this._battleCounter}`;
    }
    init(core) {
        core.events.on(this.namespace, 'battle/start', (p, output) => {
            const battleId = p.battleId ?? this._generateBattleId();
            const party = p.party.map((c) => ({ ...c, side: 'party', alive: true }));
            const enemies = p.enemies.map((c) => ({ ...c, side: 'enemy', alive: true }));
            const state = {
                battleId,
                phase: 'acting',
                turn: 1,
                party,
                enemies,
                pendingActions: [],
            };
            this._battles.set(battleId, state);
            output.battleId = battleId;
            output.state = state;
            core.events.emitSync('battle/started', { battleId, state });
            core.events.emitSync('battle/turn:start', { battleId, turn: 1 });
        });
        core.events.on(this.namespace, 'battle/action', (p) => {
            const state = this._battles.get(p.battleId);
            if (!state || state.phase !== 'acting')
                return;
            state.pendingActions.push(p.action);
        });
        core.events.on(this.namespace, 'battle/resolve', (p, output) => {
            const state = this._battles.get(p.battleId);
            if (!state || state.phase !== 'acting')
                return;
            state.phase = 'resolving';
            const results = [];
            // Sort by AGI (faster combatants act first)
            const sorted = [...state.pendingActions].sort((a, b) => this._getStat(core, b.actorId, 'agi') - this._getStat(core, a.actorId, 'agi'));
            for (const action of sorted) {
                const result = this._resolveAction(core, state, action);
                results.push(result);
                core.events.emitSync('battle/action:resolved', {
                    battleId: p.battleId,
                    action,
                    effects: result.effects,
                    critical: result.critical,
                    defeated: result.defeated,
                });
            }
            // Remove the temporary guard modifier applied this turn by any actor
            // that chose 'guard'.  The modifier is only meant to last one turn.
            for (const action of sorted) {
                if (action.kind === 'guard') {
                    core.events.emitSync('stats/modifier:remove', {
                        actorId: action.actorId,
                        source: 'guard:temp',
                    });
                }
            }
            output.results = results;
            state.pendingActions = [];
            const allEnemiesDead = state.enemies.every((e) => !e.alive);
            const allPartyDead = state.party.every((m) => !m.alive);
            if (allEnemiesDead || allPartyDead) {
                const outcome = allEnemiesDead ? 'victory' : 'defeat';
                state.phase = 'end';
                this._battles.delete(p.battleId);
                core.events.emitSync('battle/ended', { battleId: p.battleId, outcome });
            }
            else {
                state.turn++;
                state.phase = 'acting';
                core.events.emitSync('battle/turn:start', { battleId: p.battleId, turn: state.turn });
            }
        });
        core.events.on(this.namespace, 'battle/end', (p) => {
            const state = this._battles.get(p.battleId);
            if (!state)
                return;
            state.phase = 'end';
            this._battles.delete(p.battleId);
            core.events.emitSync('battle/ended', { battleId: p.battleId, outcome: p.outcome });
        });
        core.events.on(this.namespace, 'battle/state:get', (p, output) => {
            output.state = this._battles.get(p.battleId) ?? null;
        });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
    }
    // ---------------------------------------------------------------------------
    // Internal helpers
    // ---------------------------------------------------------------------------
    _resolveAction(core, state, action) {
        const effects = {};
        let critical = false;
        let defeated = false;
        const actorAtk = this._getStat(core, action.actorId, 'atk');
        for (const targetId of action.targetIds) {
            const target = [...state.party, ...state.enemies].find((c) => c.id === targetId);
            if (!target || !target.alive)
                continue;
            let damage = 0;
            switch (action.kind) {
                case 'attack': {
                    const targetDef = this._getStat(core, targetId, 'def');
                    damage = Math.max(1, actorAtk - targetDef);
                    if (Math.random() < 0.1) {
                        damage *= 2;
                        critical = true;
                    }
                    break;
                }
                case 'guard':
                    core.events.emitSync('stats/modifier:add', {
                        actorId: action.actorId,
                        source: 'guard:temp',
                        modifiers: [{ stat: 'def', value: 2, mode: 'multiply' }],
                    });
                    break;
                case 'flee':
                    break;
                default:
                    damage = Math.max(1, actorAtk);
            }
            if (damage > 0) {
                const { output: baseOut } = core.events.emitSync('stats/base:get', { actorId: targetId });
                const currentHp = baseOut?.base?.['hp'] ?? 0;
                const newHp = Math.max(0, currentHp - damage);
                core.events.emitSync('stats/base:set', { actorId: targetId, patch: { hp: newHp } });
                effects[targetId] = { delta: -damage, stat: 'hp' };
                if (newHp <= 0 && target.alive) {
                    target.alive = false;
                    defeated = true;
                    core.events.emitSync('battle/combatant:defeated', {
                        battleId: state.battleId,
                        combatantId: targetId,
                        side: target.side,
                    });
                }
            }
        }
        return { action, effects, critical, defeated };
    }
    _getStat(core, actorId, stat) {
        const { output } = core.events.emitSync('stats/compute', { actorId });
        return output?.stats?.[stat] ?? 0;
    }
}
