// @inkshot/engine/dist/plugins/rpg/RpgMenuSystem.js
// ---------------------------------------------------------------------------
// RpgMenuSystem
// ---------------------------------------------------------------------------
/**
 * Plugin that manages the **RPG pause-menu state machine**.
 *
 * This plugin is UI-agnostic: it manages the open/closed state and the
 * current page, and emits events that a separate UI layer can react to.
 *
 * ### Default key bindings
 * - `'menu'`   action → opens the main menu
 * - `'cancel'` action → goes back / closes the menu
 *
 * ### EventBus API
 *
 * | Event                  | Params / Output                                    |
 * |------------------------|----------------------------------------------------|
 * | `rpgmenu/open`         | `RpgMenuOpenParams`                                |
 * | `rpgmenu/close`        | `{}`                                               |
 * | `rpgmenu/page:set`     | `RpgMenuPageSetParams`                             |
 * | `rpgmenu/state:get`    | `{} → RpgMenuStateGetOutput`                       |
 * | `rpgmenu/party:set`    | `RpgMenuPartySetParams`                            |
 * | `rpgmenu/opened`       | `RpgMenuOpenedParams` (notification)               |
 * | `rpgmenu/closed`       | `RpgMenuClosedParams` (notification)               |
 * | `rpgmenu/page:changed` | `RpgMenuPageChangedParams` (notification)          |
 */
export class RpgMenuSystem {
    namespace = 'rpgmenu';
    dependencies = ['game'];
    editorMeta = {
        displayName: 'RPG Menu System',
        icon: 'rpgmenu',
        description: 'Manages the main RPG status/equipment menu with page navigation.',
        events: [
            'rpgmenu/open', 'rpgmenu/close', 'rpgmenu/page:set', 'rpgmenu/party:set', 'rpgmenu/state:get',
        ],
    };
    _open = false;
    _page = 'main';
    _openAction;
    _cancelAction;
    constructor(opts = {}) {
        this._openAction = opts.openAction ?? 'menu';
        this._cancelAction = opts.cancelAction ?? 'cancel';
    }
    init(core) {
        core.events.on(this.namespace, 'rpgmenu/open', (p) => {
            this._openMenu(core, p.page ?? 'main');
        });
        core.events.on(this.namespace, 'rpgmenu/close', () => {
            this._closeMenu(core);
        });
        core.events.on(this.namespace, 'rpgmenu/page:set', (p) => {
            if (!this._open)
                return;
            const previous = this._page;
            this._page = p.page;
            core.events.emitSync('rpgmenu/page:changed', { page: this._page, previous });
        });
        core.events.on(this.namespace, 'rpgmenu/state:get', (_p, output) => {
            output.open = this._open;
            output.page = this._page;
        });
        core.events.on(this.namespace, 'rpgmenu/party:set', (_p) => {
            // Party list stored externally; acknowledged for API completeness.
        });
        // Input integration
        core.events.on(this.namespace, 'input/action:triggered', (p) => {
            if (p.state !== 'pressed')
                return;
            if (p.action === this._openAction && !this._open) {
                const { output: gameOut } = core.events.emitSync('game/state:get', {});
                if (gameOut?.state === 'playing' || gameOut?.state === undefined) {
                    this._openMenu(core, 'main');
                }
                return;
            }
            if (p.action === this._cancelAction && this._open) {
                if (this._page !== 'main') {
                    const previous = this._page;
                    this._page = 'main';
                    core.events.emitSync('rpgmenu/page:changed', { page: 'main', previous });
                }
                else {
                    this._closeMenu(core);
                }
            }
        });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
    }
    // ---------------------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------------------
    _openMenu(core, page) {
        if (this._open)
            return;
        this._open = true;
        this._page = page;
        core.events.emitSync('game/state:set', { state: 'paused' });
        core.events.emitSync('rpgmenu/opened', { page });
    }
    _closeMenu(core) {
        if (!this._open)
            return;
        this._open = false;
        core.events.emitSync('game/state:set', { state: 'playing' });
        core.events.emitSync('rpgmenu/closed', {});
    }
}
