// @inkshot/engine/dist/plugins/rpg/ShopSystem.js
// ---------------------------------------------------------------------------
// ShopSystem
// ---------------------------------------------------------------------------
/**
 * Plugin that manages **shop interactions**.
 *
 * Gold is read from / written to `VariableStoreManager` under the
 * `'player'` namespace (default), key `'gold'`.
 *
 * ### EventBus API
 *
 * | Event        | Params / Output                           |
 * |--------------|-------------------------------------------|
 * | `shop/define`| `ShopDefineParams`                        |
 * | `shop/open`  | `ShopOpenParams → ShopOpenOutput`         |
 * | `shop/buy`   | `ShopBuyParams → ShopBuyOutput`           |
 * | `shop/sell`  | `ShopSellParams → ShopSellOutput`         |
 * | `shop/close` | `ShopCloseParams`                         |
 */
export class ShopSystem {
    namespace = 'shop';
    dependencies = ['inventory'];
    editorMeta = {
        displayName: 'Shop System',
        icon: 'shop',
        description: 'Shop definitions with buy/sell workflows integrated with the inventory.',
        events: [
            'shop/define', 'shop/open', 'shop/close', 'shop/buy', 'shop/sell',
        ],
    };
    _shops = new Map();
    _sessions = new Map();
    _goldNs;
    _goldKey;
    constructor(opts = {}) {
        this._goldNs = opts.goldNamespace ?? 'player';
        this._goldKey = opts.goldKey ?? 'gold';
    }
    init(core) {
        core.events.on(this.namespace, 'shop/define', (p) => {
            this._shops.set(p.shop.id, p.shop);
        });
        core.events.on(this.namespace, 'shop/open', (p, output) => {
            const shop = this._shops.get(p.shopId);
            if (!shop)
                return;
            const stock = new Map();
            const entries = shop.entries.map((entry) => {
                const itemDef = this._getItemDef(core, entry.itemId);
                const buyPrice = entry.price ?? itemDef?.price ?? 0;
                const sellRatio = shop.sellRatio ?? 0.5;
                const sellPrice = Math.floor(buyPrice * sellRatio);
                const stockCount = entry.stock ?? null;
                stock.set(entry.itemId, stockCount);
                return { itemId: entry.itemId, price: buyPrice, sellPrice, stock: stockCount };
            });
            this._sessions.set(p.customerId, { shopId: p.shopId, customerId: p.customerId, stock });
            output.entries = entries;
            output.gold = this._getGold(core);
            core.events.emitSync('shop/opened', { shopId: p.shopId, customerId: p.customerId });
        });
        core.events.on(this.namespace, 'shop/buy', (p, output) => {
            const session = this._sessions.get(p.customerId);
            if (!session || session.shopId !== p.shopId) {
                output.success = false;
                output.reason = 'no_active_session';
                output.goldSpent = 0;
                output.newGold = this._getGold(core);
                return;
            }
            const shop = this._shops.get(p.shopId);
            const entry = shop.entries.find((e) => e.itemId === p.itemId);
            if (!entry) {
                output.success = false;
                output.reason = 'item_not_available';
                output.goldSpent = 0;
                output.newGold = this._getGold(core);
                return;
            }
            const qty = p.quantity ?? 1;
            const buyPrice = (entry.price ?? this._getItemDef(core, p.itemId)?.price ?? 0) * qty;
            const currentGold = this._getGold(core);
            if (currentGold < buyPrice) {
                output.success = false;
                output.reason = 'insufficient_gold';
                output.goldSpent = 0;
                output.newGold = currentGold;
                return;
            }
            const currentStock = session.stock.get(p.itemId);
            if (currentStock !== null && currentStock !== undefined && currentStock < qty) {
                output.success = false;
                output.reason = 'out_of_stock';
                output.goldSpent = 0;
                output.newGold = currentGold;
                return;
            }
            const newGold = currentGold - buyPrice;
            this._setGold(core, newGold);
            if (currentStock !== null && currentStock !== undefined) {
                session.stock.set(p.itemId, currentStock - qty);
            }
            core.events.emitSync('inventory/add', { actorId: p.customerId, itemId: p.itemId, quantity: qty });
            output.success = true;
            output.goldSpent = buyPrice;
            output.newGold = newGold;
            core.events.emitSync('shop/bought', {
                shopId: p.shopId, customerId: p.customerId, itemId: p.itemId, quantity: qty, goldSpent: buyPrice,
            });
        });
        core.events.on(this.namespace, 'shop/sell', (p, output) => {
            const session = this._sessions.get(p.customerId);
            if (!session || session.shopId !== p.shopId) {
                output.success = false;
                output.reason = 'no_active_session';
                output.goldGained = 0;
                output.newGold = this._getGold(core);
                return;
            }
            const shop = this._shops.get(p.shopId);
            const qty = p.quantity ?? 1;
            const entry = shop.entries.find((e) => e.itemId === p.itemId);
            const itemDef = this._getItemDef(core, p.itemId);
            const basePrice = entry?.price ?? itemDef?.price ?? 0;
            const sellPrice = Math.floor(basePrice * (shop.sellRatio ?? 0.5)) * qty;
            const { output: hasOut } = core.events.emitSync('inventory/has', { actorId: p.customerId, itemId: p.itemId, quantity: qty });
            if (!hasOut.has) {
                output.success = false;
                output.reason = 'item_not_owned';
                output.goldGained = 0;
                output.newGold = this._getGold(core);
                return;
            }
            core.events.emitSync('inventory/remove', { actorId: p.customerId, itemId: p.itemId, quantity: qty });
            const newGold = this._getGold(core) + sellPrice;
            this._setGold(core, newGold);
            output.success = true;
            output.goldGained = sellPrice;
            output.newGold = newGold;
            core.events.emitSync('shop/sold', {
                shopId: p.shopId, customerId: p.customerId, itemId: p.itemId, quantity: qty, goldGained: sellPrice,
            });
        });
        core.events.on(this.namespace, 'shop/close', (p) => {
            this._sessions.delete(p.customerId);
            core.events.emitSync('shop/closed', { shopId: p.shopId, customerId: p.customerId });
        });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
    }
    // ---------------------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------------------
    _getGold(core) {
        const { output } = core.events.emitSync('store/get', { ns: this._goldNs, key: this._goldKey });
        return typeof output?.value === 'number' ? output.value : 0;
    }
    _setGold(core, amount) {
        core.events.emitSync('store/set', { ns: this._goldNs, key: this._goldKey, value: amount });
    }
    _getItemDef(core, itemId) {
        const { output } = core.events.emitSync('inventory/item:get', { itemId });
        return output?.def;
    }
}
