// @inkshot/engine/dist/plugins/rpg/StatsSystem.js
// ---------------------------------------------------------------------------
// StatsSystem
// ---------------------------------------------------------------------------
/**
 * Plugin that manages **actor and enemy statistics**.
 *
 * The system maintains per-actor base stats, equipment / buff modifiers, and
 * status effects (poison, haste, etc.).  Fully computed stats are available
 * via `stats/compute`.
 *
 * ### EventBus API
 *
 * | Event                     | Params / Output                                   |
 * |---------------------------|---------------------------------------------------|
 * | `stats/profile:define`    | `StatsProfileDefineParams`                        |
 * | `stats/status:define`     | `{ effect: StatusEffectDef }`                     |
 * | `stats/base:set`          | `StatsBaseSetParams`                              |
 * | `stats/base:get`          | `StatsBaseGetParams → StatsBaseGetOutput`         |
 * | `stats/modifier:add`      | `StatsModifierAddParams`                          |
 * | `stats/modifier:remove`   | `StatsModifierRemoveParams`                       |
 * | `stats/compute`           | `StatsComputeParams → StatsComputeOutput`         |
 * | `stats/status:apply`      | `StatsStatusApplyParams`                          |
 * | `stats/status:remove`     | `StatsStatusRemoveParams`                         |
 * | `stats/status:list`       | `StatsStatusListParams → StatsStatusListOutput`   |
 * | `stats/changed`           | `StatsChangedParams` (notification)               |
 * | `stats/status:applied`    | `StatsStatusAppliedParams` (notification)         |
 * | `stats/status:expired`    | `StatsStatusExpiredParams` (notification)         |
 */
export class StatsSystem {
    namespace = 'stats';
    editorMeta = {
        displayName: 'Stats System',
        icon: 'stats',
        description: 'Manages actor stat profiles, modifiers, and status effects.',
        events: [
            'stats/profile:define', 'stats/base:set', 'stats/base:get',
            'stats/modifier:add', 'stats/modifier:remove', 'stats/compute',
            'stats/status:define', 'stats/status:apply', 'stats/status:remove', 'stats/status:list',
        ],
    };
    /** Registered stat profile definitions (by id). */
    _profiles = new Map();
    /** Registered status effect definitions (by id). */
    _effectDefs = new Map();
    /** Per-actor base stats. */
    _bases = new Map();
    /**
     * Per-actor modifier map:  actorId → sourceKey → modifiers[].
     * Tracks equipment and buff sources independently so they can be removed
     * cleanly.
     */
    _modifiers = new Map();
    /** Per-actor active status effects. */
    _statuses = new Map();
    init(core) {
        core.events.on(this.namespace, 'stats/profile:define', (p) => {
            this._profiles.set(p.profile.id, p.profile);
        });
        core.events.on(this.namespace, 'stats/status:define', (p) => {
            this._effectDefs.set(p.effect.id, p.effect);
        });
        core.events.on(this.namespace, 'stats/base:set', (p) => {
            const existing = this._bases.get(p.actorId) ?? {};
            this._bases.set(p.actorId, { ...existing, ...p.patch });
            this._emitChanged(core, p.actorId);
        });
        core.events.on(this.namespace, 'stats/base:get', (p, output) => {
            output.base = { ...(this._bases.get(p.actorId) ?? {}) };
        });
        core.events.on(this.namespace, 'stats/modifier:add', (p) => {
            if (!this._modifiers.has(p.actorId)) {
                this._modifiers.set(p.actorId, new Map());
            }
            this._modifiers.get(p.actorId).set(p.source, [...p.modifiers]);
            this._emitChanged(core, p.actorId);
        });
        core.events.on(this.namespace, 'stats/modifier:remove', (p) => {
            this._modifiers.get(p.actorId)?.delete(p.source);
            this._emitChanged(core, p.actorId);
        });
        core.events.on(this.namespace, 'stats/compute', (p, output) => {
            output.stats = this._computeStats(p.actorId);
        });
        core.events.on(this.namespace, 'stats/status:apply', (p) => {
            this._applyStatus(core, p.actorId, p.effectId);
        });
        core.events.on(this.namespace, 'stats/status:remove', (p) => {
            this._removeStatus(core, p.actorId, p.effectId);
        });
        core.events.on(this.namespace, 'stats/status:list', (p, output) => {
            output.effectIds = [...(this._statuses.get(p.actorId)?.keys() ?? [])];
        });
        // Re-broadcast stats/changed when the actor's level changes so that any
        // system that derives values from both level and stats stays in sync.
        core.events.on(this.namespace, 'stats/base:level-changed', (p) => {
            this._emitChanged(core, p.actorId);
        });
        // Drive status effect timers and tick damage from the update loop.
        core.events.on(this.namespace, 'core/update', (p) => {
            this._tickStatuses(core, p.dt);
        });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
    }
    // ---------------------------------------------------------------------------
    // Internal helpers
    // ---------------------------------------------------------------------------
    _computeStats(actorId) {
        const base = { ...(this._bases.get(actorId) ?? {}) };
        // Collect all add-modifiers first, then multiply.
        const adds = {};
        const mults = {};
        const collectMods = (mods) => {
            for (const m of mods) {
                const mode = m.mode ?? 'add';
                if (mode === 'add') {
                    adds[m.stat] = (adds[m.stat] ?? 0) + m.value;
                }
                else {
                    mults[m.stat] = (mults[m.stat] ?? 1) * m.value;
                }
            }
        };
        // Equipment / buff modifiers
        for (const srcMap of (this._modifiers.get(actorId) ?? new Map()).values()) {
            collectMods(srcMap);
        }
        // Status effect modifiers
        for (const status of (this._statuses.get(actorId) ?? new Map()).values()) {
            collectMods(status.def.modifiers);
        }
        const result = { ...base };
        for (const [stat, val] of Object.entries(adds)) {
            result[stat] = (result[stat] ?? 0) + val;
        }
        for (const [stat, mul] of Object.entries(mults)) {
            result[stat] = (result[stat] ?? 0) * mul;
        }
        return result;
    }
    _emitChanged(core, actorId) {
        const stats = this._computeStats(actorId);
        core.events.emitSync('stats/changed', { actorId, stats });
    }
    _applyStatus(core, actorId, effectId) {
        const def = this._effectDefs.get(effectId);
        if (!def)
            return;
        if (!this._statuses.has(actorId)) {
            this._statuses.set(actorId, new Map());
        }
        const remaining = def.duration !== undefined ? def.duration : Infinity;
        this._statuses.get(actorId).set(effectId, { def, remaining, tickAccum: 0 });
        this._emitChanged(core, actorId);
        core.events.emitSync('stats/status:applied', { actorId, effectId, effect: def });
    }
    _removeStatus(core, actorId, effectId) {
        if (!this._statuses.get(actorId)?.has(effectId))
            return;
        this._statuses.get(actorId).delete(effectId);
        this._emitChanged(core, actorId);
        core.events.emitSync('stats/status:expired', { actorId, effectId });
    }
    _tickStatuses(core, deltaMs) {
        for (const [actorId, statusMap] of this._statuses) {
            const toRemove = [];
            for (const [effectId, status] of statusMap) {
                // Tick damage
                if (status.def.tickDamage !== undefined) {
                    const tickInterval = status.def.tickMs ?? 1000;
                    status.tickAccum += deltaMs;
                    while (status.tickAccum >= tickInterval) {
                        status.tickAccum -= tickInterval;
                        const base = this._bases.get(actorId) ?? {};
                        const current = base['hp'] ?? 0;
                        const newHp = Math.max(0, current - status.def.tickDamage);
                        this._bases.set(actorId, { ...base, hp: newHp });
                        this._emitChanged(core, actorId);
                    }
                }
                // Duration countdown
                if (status.remaining !== Infinity) {
                    status.remaining -= deltaMs;
                    if (status.remaining <= 0) {
                        toRemove.push(effectId);
                    }
                }
            }
            for (const id of toRemove) {
                this._removeStatus(core, actorId, id);
            }
        }
    }
}
