// @inkshot/engine/dist/rendering/Camera.js
// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------
function lerp(a, b, t) {
    return a + (b - a) * t;
}
function clamp(value, min, max) {
    return Math.max(min, Math.min(max, value));
}
// ---------------------------------------------------------------------------
// Camera
// ---------------------------------------------------------------------------
/**
 * 2D camera that manipulates a world container's transform to simulate
 * viewport movement, zooming, rotation, and screen shake.
 *
 * The camera works by moving the **world layer** in the opposite direction
 * of the logical camera position, so the coordinate system for game objects
 * remains unchanged.
 *
 * ### Quick start
 * ```ts
 * const camera = new Camera(core, renderer.getLayer('world'), {
 *   viewportWidth: 1280,
 *   viewportHeight: 720,
 * });
 *
 * camera.follow(player, { lerp: 0.08, deadzone: { width: 64, height: 64 } });
 * camera.setBounds({ x: 0, y: 0, width: 4096, height: 4096 });
 * ```
 *
 * ### EventBus API
 * | Event               | Params                   | Description              |
 * |---------------------|--------------------------|--------------------------|
 * | `camera/follow`     | `CameraFollowParams`     | Set or clear follow      |
 * | `camera/shake`      | `CameraShakeParams`      | Trigger screen shake     |
 * | `camera/move`       | `CameraMoveParams`       | Move to / offset by      |
 * | `camera/zoom`       | `CameraZoomParams`       | Set zoom level           |
 * | `camera/state`      | `{}`                     | Read current state (output) |
 */
export class Camera {
    _core;
    _worldContainer;
    // Viewport dimensions (screen space).
    _viewportW;
    _viewportH;
    // Logical camera position (world-space centre).
    _x = 0;
    _y = 0;
    _zoom = 1;
    _rotation = 0;
    // Follow target.
    _target = null;
    _followLerp = 0.1;
    _deadzone = null;
    _followOffset = { x: 0, y: 0 };
    // World bounds (null = unconstrained).
    _bounds = null;
    // Screen shake state.
    _shakeIntensity = 0;
    _shakeDuration = 0;
    _shakeElapsed = 0;
    _shakeDecay = 'exponential';
    _shakeOffsetX = 0;
    _shakeOffsetY = 0;
    constructor(core, worldContainer, opts) {
        this._core = core;
        this._worldContainer = worldContainer;
        this._viewportW = opts.viewportWidth;
        this._viewportH = opts.viewportHeight;
        // Subscribe to fixed update (follow + shake advance).
        core.events.on('camera', 'core/update', this._onUpdate, { priority: 0 });
        // Subscribe to pre-render (apply transform with interpolation-friendly timing).
        core.events.on('camera', 'renderer/pre-render', this._onPreRender, { priority: 0 });
        // ── EventBus public API ──────────────────────────────────────────────
        core.events.on('camera', 'camera/follow', (p) => {
            if (p.target) {
                this.follow(p.target, p.options);
            }
            else {
                this.unfollow();
            }
        });
        core.events.on('camera', 'camera/shake', (p) => {
            this.shake(p);
        });
        core.events.on('camera', 'camera/move', (p) => {
            if (p.relative) {
                this.moveBy(p.x, p.y);
            }
            else {
                this.moveTo(p.x, p.y);
            }
        });
        core.events.on('camera', 'camera/zoom', (p) => {
            this.setZoom(p.zoom);
        });
        core.events.on('camera', 'camera/state', (_p, output) => {
            output.x = this._x;
            output.y = this._y;
            output.zoom = this._zoom;
            output.rotation = this._rotation;
            output.viewportWidth = this._viewportW;
            output.viewportHeight = this._viewportH;
        });
    }
    // ---------------------------------------------------------------------------
    // Public API — position
    // ---------------------------------------------------------------------------
    get x() { return this._x; }
    get y() { return this._y; }
    moveTo(x, y) {
        this._x = x;
        this._y = y;
        this._clampToBounds();
    }
    moveBy(dx, dy) {
        this._x += dx;
        this._y += dy;
        this._clampToBounds();
    }
    // ---------------------------------------------------------------------------
    // Public API — zoom & rotation
    // ---------------------------------------------------------------------------
    get zoom() { return this._zoom; }
    setZoom(value) {
        this._zoom = Math.max(0.01, value);
        this._clampToBounds();
    }
    get rotation() { return this._rotation; }
    setRotation(radians) {
        this._rotation = radians;
    }
    // ---------------------------------------------------------------------------
    // Public API — viewport
    // ---------------------------------------------------------------------------
    get viewportWidth() { return this._viewportW; }
    get viewportHeight() { return this._viewportH; }
    setViewport(width, height) {
        this._viewportW = width;
        this._viewportH = height;
        this._clampToBounds();
    }
    // ---------------------------------------------------------------------------
    // Public API — follow
    // ---------------------------------------------------------------------------
    follow(target, options) {
        this._target = target;
        this._followLerp = options?.lerp ?? 0.1;
        this._deadzone = options?.deadzone ?? null;
        this._followOffset = options?.offset ?? { x: 0, y: 0 };
    }
    unfollow() {
        this._target = null;
    }
    // ---------------------------------------------------------------------------
    // Public API — bounds
    // ---------------------------------------------------------------------------
    setBounds(rect) {
        this._bounds = rect;
        if (rect)
            this._clampToBounds();
    }
    get bounds() {
        return this._bounds;
    }
    // ---------------------------------------------------------------------------
    // Public API — shake
    // ---------------------------------------------------------------------------
    shake(opts) {
        this._shakeIntensity = opts.intensity;
        this._shakeDuration = opts.duration;
        this._shakeElapsed = 0;
        this._shakeDecay = opts.decay ?? 'exponential';
    }
    // ---------------------------------------------------------------------------
    // Lifecycle
    // ---------------------------------------------------------------------------
    destroy() {
        this._core.events.removeNamespace('camera');
    }
    // ---------------------------------------------------------------------------
    // Private — game logic (fixed step)
    // ---------------------------------------------------------------------------
    _onUpdate = (params) => {
        // ── Follow target ────────────────────────────────────────────────────
        if (this._target) {
            const goalX = this._target.x + this._followOffset.x;
            const goalY = this._target.y + this._followOffset.y;
            if (this._deadzone) {
                const halfW = this._deadzone.width / 2;
                const halfH = this._deadzone.height / 2;
                const dx = goalX - this._x;
                const dy = goalY - this._y;
                // Only move when target exits the deadzone.
                if (Math.abs(dx) > halfW) {
                    const edge = dx > 0 ? goalX - halfW : goalX + halfW;
                    this._x = lerp(this._x, edge, this._followLerp);
                }
                if (Math.abs(dy) > halfH) {
                    const edge = dy > 0 ? goalY - halfH : goalY + halfH;
                    this._y = lerp(this._y, edge, this._followLerp);
                }
            }
            else {
                this._x = lerp(this._x, goalX, this._followLerp);
                this._y = lerp(this._y, goalY, this._followLerp);
            }
            this._clampToBounds();
        }
        // ── Shake advance ────────────────────────────────────────────────────
        if (this._shakeDuration > 0 && this._shakeElapsed < this._shakeDuration) {
            this._shakeElapsed += params.dt;
            const progress = clamp(this._shakeElapsed / this._shakeDuration, 0, 1);
            const factor = this._shakeDecay === 'exponential'
                ? 1 - progress * progress
                : 1 - progress;
            const magnitude = this._shakeIntensity * factor;
            this._shakeOffsetX = (Math.random() * 2 - 1) * magnitude;
            this._shakeOffsetY = (Math.random() * 2 - 1) * magnitude;
            if (this._shakeElapsed >= this._shakeDuration) {
                this._shakeOffsetX = 0;
                this._shakeOffsetY = 0;
            }
        }
    };
    // ---------------------------------------------------------------------------
    // Private — apply transform (render frame)
    // ---------------------------------------------------------------------------
    _onPreRender = (_params) => {
        const hw = this._viewportW / 2;
        const hh = this._viewportH / 2;
        // Translate so that the camera centre maps to the viewport centre,
        // then apply zoom and rotation around the centre.
        this._worldContainer.x = -this._x * this._zoom + hw + this._shakeOffsetX;
        this._worldContainer.y = -this._y * this._zoom + hh + this._shakeOffsetY;
        this._worldContainer.scale.set(this._zoom);
        this._worldContainer.rotation = -this._rotation;
        // Set the pivot to the camera position so rotation & zoom are relative to
        // the centre of the viewport.
        this._worldContainer.pivot.set(0, 0);
    };
    // ---------------------------------------------------------------------------
    // Private — bounds clamping
    // ---------------------------------------------------------------------------
    _clampToBounds() {
        if (!this._bounds)
            return;
        const halfViewW = (this._viewportW / this._zoom) / 2;
        const halfViewH = (this._viewportH / this._zoom) / 2;
        const minX = this._bounds.x + halfViewW;
        const maxX = this._bounds.x + this._bounds.width - halfViewW;
        const minY = this._bounds.y + halfViewH;
        const maxY = this._bounds.y + this._bounds.height - halfViewH;
        // If the viewport is larger than the bounds, centre on the bounds.
        if (minX >= maxX) {
            this._x = this._bounds.x + this._bounds.width / 2;
        }
        else {
            this._x = clamp(this._x, minX, maxX);
        }
        if (minY >= maxY) {
            this._y = this._bounds.y + this._bounds.height / 2;
        }
        else {
            this._y = clamp(this._y, minY, maxY);
        }
    }
}
