// @inkshot/engine/dist/plugins/world/FogOfWarPlugin.js
import { Graphics } from 'pixi.js';
// ---------------------------------------------------------------------------
// Internal tile state encoding
// ---------------------------------------------------------------------------
const FOG_UNEXPLORED = 0;
const FOG_EXPLORED = 1;
const FOG_VISIBLE = 2;
// ---------------------------------------------------------------------------
// FogOfWarPlugin
// ---------------------------------------------------------------------------
/**
 * Plugin that provides a **fog-of-war** system based on a per-tile visibility grid.
 *
 * Each tile starts as `'unexplored'`.  When `fog/update` is called with an
 * observer position and radius, tiles within the radius become `'visible'`.
 * On the next `fog/update` call, previously visible tiles revert to
 * `'explored'` (dimly shown) unless they are still in range.
 *
 * The plugin renders the fog as a Pixi `Graphics` overlay on the `'world'`
 * layer:
 *  - Unexplored tiles → fully opaque dark squares.
 *  - Explored tiles   → semi-transparent dark squares.
 *  - Visible tiles    → no overlay (fully transparent).
 *
 * ### Quick start
 * ```ts
 * import { createEngine, FogOfWarPlugin } from 'inkshot-engine';
 *
 * const fog = new FogOfWarPlugin();
 * const { core } = await createEngine({ plugins: [fog] });
 *
 * core.events.emitSync('fog/init', {
 *   config: { mapWidth: 50, mapHeight: 40, tileWidth: 16, tileHeight: 16 },
 * });
 *
 * // Called every tick with the player's world position.
 * core.events.emitSync('fog/update', { x: player.x, y: player.y, radius: 6 });
 * ```
 *
 * ### EventBus API
 *
 * | Event              | Params / Output                                       |
 * |--------------------|-------------------------------------------------------|
 * | `fog/init`         | `FogInitParams`                                       |
 * | `fog/update`       | `FogUpdateParams`                                     |
 * | `fog/reveal`       | `FogRevealParams`                                     |
 * | `fog/get-tile`     | `FogGetTileParams → FogGetTileOutput`                 |
 * | `fog/clear`        | `{}`                                                  |
 * | `fog/state`        | `{} → FogStateOutput`                                 |
 * | `fog/tile:revealed`| `FogTileRevealedParams` (notification)                |
 */
export class FogOfWarPlugin {
    namespace = 'fog';
    editorMeta = {
        displayName: 'Fog of War Plugin',
        icon: 'fog',
        description: 'Renders a fog-of-war overlay with explored and visible tile states.',
        events: [
            'fog/init', 'fog/update', 'fog/reveal', 'fog/clear',
            'fog/get-tile', 'fog/state',
        ],
    };
    _config = null;
    /** Flat row-major array: 0=unexplored, 1=explored, 2=visible. */
    _grid = null;
    _gfx = null;
    _layer = null;
    // ---------------------------------------------------------------------------
    // EnginePlugin lifecycle
    // ---------------------------------------------------------------------------
    init(core) {
        // ── EventBus API ─────────────────────────────────────────────────────
        core.events.on('fog', 'fog/init', (p) => {
            this._initGrid(p.config, core);
        });
        core.events.on('fog', 'fog/update', (p) => {
            if (!this._config || !this._grid)
                return;
            this._updateVisibility(p.x, p.y, p.radius, core);
        });
        core.events.on('fog', 'fog/reveal', (p) => {
            if (!this._config || !this._grid)
                return;
            const targetState = p.state ?? 'explored';
            const stateCode = _stateToCode(targetState);
            const { mapWidth, mapHeight } = this._config;
            const maxRow = Math.min(p.row + p.height, mapHeight);
            const maxCol = Math.min(p.col + p.width, mapWidth);
            for (let r = Math.max(0, p.row); r < maxRow; r++) {
                for (let c = Math.max(0, p.col); c < maxCol; c++) {
                    const idx = r * mapWidth + c;
                    const prev = this._grid[idx];
                    if (stateCode > prev) {
                        if (prev === FOG_UNEXPLORED) {
                            core.events.emitSync('fog/tile:revealed', { col: c, row: r });
                        }
                        this._grid[idx] = stateCode;
                    }
                }
            }
        });
        core.events.on('fog', 'fog/get-tile', (p, output) => {
            output.state = this._getTileState(p.col, p.row);
        });
        core.events.on('fog', 'fog/clear', () => {
            if (this._grid)
                this._grid.fill(FOG_UNEXPLORED);
        });
        core.events.on('fog', 'fog/state', (_p, output) => {
            output.config = this._config ? { ...this._config } : null;
            if (this._grid) {
                let explored = 0;
                let visible = 0;
                for (const v of this._grid) {
                    if (v === FOG_EXPLORED)
                        explored++;
                    else if (v === FOG_VISIBLE)
                        visible++;
                }
                output.total = this._grid.length;
                output.explored = explored;
                output.visible = visible;
            }
            else {
                output.total = 0;
                output.explored = 0;
                output.visible = 0;
            }
        });
        // ── Redraw fog every render frame ────────────────────────────────────
        core.events.on('fog', 'renderer/pre-render', (_p) => {
            this._draw();
        });
    }
    destroy(core) {
        if (this._gfx) {
            this._gfx.destroy();
            this._gfx = null;
        }
        this._grid = null;
        this._config = null;
        core.events.removeNamespace('fog');
    }
    // ---------------------------------------------------------------------------
    // Direct accessors (pull API)
    // ---------------------------------------------------------------------------
    getTileState(col, row) {
        return this._getTileState(col, row);
    }
    // ---------------------------------------------------------------------------
    // Private — grid initialisation
    // ---------------------------------------------------------------------------
    _initGrid(config, core) {
        this._config = { ...config };
        this._grid = new Uint8Array(config.mapWidth * config.mapHeight).fill(FOG_UNEXPLORED);
        // Tear down and recreate the graphics layer.
        if (this._gfx) {
            this._gfx.destroy();
            this._gfx = null;
        }
        const { output: layerResult } = core.events.emitSync('renderer/layer:create', { name: 'fog', zIndex: 75 });
        this._layer = layerResult.layer ?? null;
        if (this._layer) {
            this._gfx = new Graphics();
            this._layer.addChild(this._gfx);
        }
    }
    // ---------------------------------------------------------------------------
    // Private — visibility update
    // ---------------------------------------------------------------------------
    _updateVisibility(worldX, worldY, radiusTiles, core) {
        const cfg = this._config;
        const grid = this._grid;
        // Revert previously visible tiles to explored.
        for (let i = 0; i < grid.length; i++) {
            if (grid[i] === FOG_VISIBLE)
                grid[i] = FOG_EXPLORED;
        }
        // Compute the observer's tile position.
        const centerCol = Math.floor(worldX / cfg.tileWidth);
        const centerRow = Math.floor(worldY / cfg.tileHeight);
        const r = Math.ceil(radiusTiles);
        for (let dr = -r; dr <= r; dr++) {
            for (let dc = -r; dc <= r; dc++) {
                // Circular mask.
                if (dc * dc + dr * dr > radiusTiles * radiusTiles)
                    continue;
                const col = centerCol + dc;
                const row = centerRow + dr;
                if (col < 0 || col >= cfg.mapWidth || row < 0 || row >= cfg.mapHeight)
                    continue;
                const idx = row * cfg.mapWidth + col;
                const prev = grid[idx];
                if (prev === FOG_UNEXPLORED) {
                    core.events.emitSync('fog/tile:revealed', { col, row });
                }
                grid[idx] = FOG_VISIBLE;
            }
        }
    }
    // ---------------------------------------------------------------------------
    // Private — rendering
    // ---------------------------------------------------------------------------
    _draw() {
        if (!this._gfx || !this._config || !this._grid)
            return;
        const cfg = this._config;
        const gfx = this._gfx;
        const unexploredAlpha = cfg.unexploredAlpha ?? 1.0;
        const exploredAlpha = cfg.exploredAlpha ?? 0.5;
        const unexploredColor = cfg.unexploredColor ?? 0x000000;
        const exploredColor = cfg.exploredColor ?? 0x000000;
        gfx.clear();
        for (let row = 0; row < cfg.mapHeight; row++) {
            for (let col = 0; col < cfg.mapWidth; col++) {
                const state = this._grid[row * cfg.mapWidth + col];
                if (state === FOG_VISIBLE)
                    continue; // no overlay for visible tiles
                const x = col * cfg.tileWidth;
                const y = row * cfg.tileHeight;
                if (state === FOG_UNEXPLORED) {
                    gfx.rect(x, y, cfg.tileWidth, cfg.tileHeight)
                        .fill({ color: unexploredColor, alpha: unexploredAlpha });
                }
                else {
                    // FOG_EXPLORED
                    gfx.rect(x, y, cfg.tileWidth, cfg.tileHeight)
                        .fill({ color: exploredColor, alpha: exploredAlpha });
                }
            }
        }
    }
    // ---------------------------------------------------------------------------
    // Private — helpers
    // ---------------------------------------------------------------------------
    _getTileState(col, row) {
        if (!this._config || !this._grid)
            return 'unexplored';
        const { mapWidth, mapHeight } = this._config;
        if (col < 0 || col >= mapWidth || row < 0 || row >= mapHeight)
            return 'unexplored';
        return _codeToState(this._grid[row * mapWidth + col]);
    }
}
// ---------------------------------------------------------------------------
// Module-level helpers
// ---------------------------------------------------------------------------
function _stateToCode(state) {
    if (state === 'explored')
        return FOG_EXPLORED;
    if (state === 'visible')
        return FOG_VISIBLE;
    return FOG_UNEXPLORED;
}
function _codeToState(code) {
    if (code === FOG_EXPLORED)
        return 'explored';
    if (code === FOG_VISIBLE)
        return 'visible';
    return 'unexplored';
}
