// @inkshot/engine/dist/plugins/world/ParallaxPlugin.js
import { Container } from 'pixi.js';
// ---------------------------------------------------------------------------
// ID generation
// ---------------------------------------------------------------------------
let _nextParallaxId = 0;
function generateParallaxId() {
    return `parallax_${++_nextParallaxId}`;
}
// ---------------------------------------------------------------------------
// ParallaxPlugin
// ---------------------------------------------------------------------------
/**
 * Plugin that provides **multi-layer parallax scrolling**.
 *
 * Each registered layer has a configurable `factorX` / `factorY` that
 * controls how fast it moves relative to the camera.  Layers with a factor
 * close to `0` appear to be in the far background; layers with a factor
 * greater than `1` rush past the camera as foreground elements.
 *
 * The plugin listens to `renderer/pre-render` and computes the correct
 * offset for every layer based on the current camera position retrieved
 * from `camera/state`.
 *
 * ### Quick start
 * ```ts
 * import { createEngine, ParallaxPlugin } from 'inkshot-engine';
 *
 * const parallax = new ParallaxPlugin();
 * const { core } = await createEngine({ plugins: [parallax] });
 *
 * // Add two background layers
 * core.events.emitSync('parallax/layer:add', { factorX: 0.2, factorY: 0 });
 * core.events.emitSync('parallax/layer:add', { factorX: 0.5, factorY: 0 });
 * ```
 *
 * ### EventBus API
 *
 * | Event                   | Params / Output                                          |
 * |-------------------------|----------------------------------------------------------|
 * | `parallax/layer:add`    | `ParallaxLayerAddParams → ParallaxLayerAddOutput`        |
 * | `parallax/layer:remove` | `ParallaxLayerRemoveParams`                              |
 * | `parallax/layer:update` | `ParallaxLayerUpdateParams`                              |
 * | `parallax/layer:get`    | `ParallaxLayerGetParams → ParallaxLayerGetOutput`        |
 * | `parallax/layers`       | `{} → ParallaxLayersOutput`                              |
 */
export class ParallaxPlugin {
    namespace = 'parallax';
    editorMeta = {
        displayName: 'Parallax Plugin',
        icon: 'parallax',
        description: 'Manages multi-layer parallax scrolling backgrounds driven by the camera.',
        events: [
            'parallax/layer:add', 'parallax/layer:remove', 'parallax/layer:update',
            'parallax/layer:get', 'parallax/layers',
        ],
    };
    _opts;
    _layers = new Map();
    constructor(opts = {}) {
        this._opts = {
            parentLayer: opts.parentLayer ?? 'world',
        };
    }
    // ---------------------------------------------------------------------------
    // EnginePlugin lifecycle
    // ---------------------------------------------------------------------------
    init(core) {
        // ── EventBus API ─────────────────────────────────────────────────────
        core.events.on('parallax', 'parallax/layer:add', (p, output) => {
            const id = p.id ?? generateParallaxId();
            const factorX = p.factorX ?? 0.5;
            const factorY = p.factorY ?? factorX;
            const def = {
                id,
                factorX,
                factorY,
                originX: p.originX ?? 0,
                originY: p.originY ?? 0,
            };
            // Create the Pixi Container and add it to the parent layer.
            const container = new Container();
            const { output: parentLayerResult } = core.events.emitSync('renderer/layer', { name: this._opts.parentLayer });
            if (parentLayerResult.layer) {
                parentLayerResult.layer.addChild(container);
            }
            this._layers.set(id, { def, container });
            output.id = id;
        });
        core.events.on('parallax', 'parallax/layer:remove', (p) => {
            const state = this._layers.get(p.id);
            if (!state)
                return;
            state.container.destroy?.();
            this._layers.delete(p.id);
        });
        core.events.on('parallax', 'parallax/layer:update', (p) => {
            const state = this._layers.get(p.id);
            if (!state)
                return;
            const def = state.def;
            if (p.factorX !== undefined)
                def.factorX = p.factorX;
            if (p.factorY !== undefined)
                def.factorY = p.factorY;
            if (p.originX !== undefined)
                def.originX = p.originX;
            if (p.originY !== undefined)
                def.originY = p.originY;
        });
        core.events.on('parallax', 'parallax/layer:get', (p, output) => {
            const state = this._layers.get(p.id);
            output.layer = state ? { ...state.def } : null;
        });
        core.events.on('parallax', 'parallax/layers', (_p, output) => {
            output.layers = Array.from(this._layers.values()).map((s) => ({ ...s.def }));
        });
        // ── Update layer positions every render frame ─────────────────────────
        core.events.on('parallax', 'renderer/pre-render', (_p) => {
            this._applyOffsets(core);
        });
    }
    destroy(core) {
        for (const state of this._layers.values()) {
            state.container.destroy?.();
        }
        this._layers.clear();
        core.events.removeNamespace('parallax');
    }
    // ---------------------------------------------------------------------------
    // Direct accessors (pull API)
    // ---------------------------------------------------------------------------
    /**
     * Return a snapshot of all registered parallax layer definitions.
     * Useful when you need to enumerate layers without going through the event bus.
     */
    getLayers() {
        return Array.from(this._layers.values()).map((s) => ({ ...s.def }));
    }
    /**
     * Return the live Pixi container for a given layer id, or `null` if not found.
     * This container should be used to add display objects to the layer.
     */
    getContainer(id) {
        return this._layers.get(id)?.container ?? null;
    }
    // ---------------------------------------------------------------------------
    // Private
    // ---------------------------------------------------------------------------
    _applyOffsets(core) {
        if (this._layers.size === 0)
            return;
        const { output: camOut } = core.events.emitSync('camera/state', {});
        const camX = camOut.x ?? 0;
        const camY = camOut.y ?? 0;
        for (const { def, container } of this._layers.values()) {
            // Offset = -(cameraPos * factor) so the layer shifts in the correct direction.
            container.x = def.originX - camX * def.factorX;
            container.y = def.originY - camY * def.factorY;
        }
    }
}
