// @inkshot/engine/dist/plugins/world/PathfindingManager.js
import { WorkerBridge } from '../../core/WorkerBridge.js';
// ---------------------------------------------------------------------------
// Internal types
// ---------------------------------------------------------------------------
/** Cost value that marks a cell as completely impassable. */
const IMPASSABLE = Infinity;
/** Maximum number of cached A* results before the oldest entry is evicted. */
const CACHE_MAX_SIZE = 512;
/** Maximum number of BFS expansions in `_findNearestPassable`. */
const NEAREST_MAX_EXPANSIONS = 1_000;
// ---------------------------------------------------------------------------
// Min-heap (binary heap) for the A* open set
// ---------------------------------------------------------------------------
class MinHeap {
    _data = [];
    get size() {
        return this._data.length;
    }
    push(node) {
        this._data.push(node);
        this._bubbleUp(this._data.length - 1);
    }
    pop() {
        if (this._data.length === 0)
            return undefined;
        const top = this._data[0];
        const last = this._data.pop();
        if (this._data.length > 0) {
            this._data[0] = last;
            this._siftDown(0);
        }
        return top;
    }
    _bubbleUp(i) {
        while (i > 0) {
            const parent = (i - 1) >> 1;
            if (this._data[parent].f <= this._data[i].f)
                break;
            [this._data[parent], this._data[i]] = [this._data[i], this._data[parent]];
            i = parent;
        }
    }
    _siftDown(i) {
        const n = this._data.length;
        while (true) {
            let smallest = i;
            const left = 2 * i + 1;
            const right = 2 * i + 2;
            if (left < n && this._data[left].f < this._data[smallest].f)
                smallest = left;
            if (right < n && this._data[right].f < this._data[smallest].f)
                smallest = right;
            if (smallest === i)
                break;
            [this._data[smallest], this._data[i]] = [this._data[i], this._data[smallest]];
            i = smallest;
        }
    }
}
/**
 * Built-in plugin providing A* pathfinding on top of the tile collision map.
 *
 * ### Design
 * - **Static obstacles** — on `physics/tilemap:set` (or directly when a
 *   tilemap is loaded via `TilemapManager`) the manager rebuilds an internal
 *   cost grid.  Tiles whose collision shape is `'solid'` (or any non-`'empty'`
 *   shape) default to {@link IMPASSABLE}; all other cells default to cost `1`.
 *   Costs can be overridden per tile-value with `pathfinding/weight:set`.
 * - **Dynamic obstacles** — when `includeDynamicObstacles: true` is set on a
 *   `pathfinding/find` call, BODY-layer entities are queried via `entity/query`
 *   and their tile cells are treated as impassable for that query only.  These
 *   results are **not** cached.
 * - **Path cache** — static searches (no dynamic obstacles) are keyed by
 *   `"fromRow,fromCol→toRow,toCol"`.  The cache is automatically invalidated
 *   whenever `physics/tilemap:set` fires.  Manual invalidation is available
 *   via `pathfinding/cache:clear`.
 * - **A\* heuristic** — Chebyshev distance (correct for 8-directional grids),
 *   or Manhattan distance when configured to 4-directional movement.
 *
 * ### EventBus API
 * | Event                      | Async? | Description                                  |
 * |----------------------------|--------|----------------------------------------------|
 * | `pathfinding/find`         | ✗ sync | Run A* from `from` to `to` (world px)        |
 * | `pathfinding/find:async`   | ✓ async| Run A* — offloads to Worker when `workerUrl` is set |
 * | `pathfinding/weight:set`   | ✗ sync | Override movement cost for a tile value      |
 * | `pathfinding/cache:clear`  | ✗ sync | Manually clear the path cache                |
 *
 * **Automatic invalidation** — the cost grid and path cache are updated
 * automatically on `physics/tilemap:set` (full reload) and on
 * `tilemap/set-tile` (single-cell O(1) update).
 *
 * ### Usage
 * ```ts
 * import { createEngine, PathfindingManager } from 'inkshot-engine';
 *
 * const { core } = await createEngine({
 *   plugins: [
 *     new KinematicPhysicsAdapter(),
 *     new EntityManager(),
 *     new PathfindingManager(),
 *   ],
 * });
 *
 * // Synchronous (main thread):
 * const { output } = core.events.emitSync<PathfindingFindParams, PathfindingFindOutput>(
 *   'pathfinding/find',
 *   { from: { x: 32, y: 32 }, to: { x: 160, y: 96 } },
 * );
 * if (output.found) console.log('Path:', output.path);
 *
 * // Async / Worker-offloaded:
 * const { output } = await core.events.emit<PathfindingFindParams, PathfindingFindOutput>(
 *   'pathfinding/find:async',
 *   { from: { x: 32, y: 32 }, to: { x: 160, y: 96 } },
 * );
 * if (output.found) console.log('Path (async):', output.path);
 * ```
 */
export class PathfindingManager {
    namespace = 'pathfinding';
    /** Must be initialised after EntityManager (entity queries used at runtime). */
    dependencies = ['entityManager'];
    editorMeta = {
        displayName: 'Pathfinding Manager',
        icon: 'pathfinding',
        description: 'Grid-based A* pathfinding with weight maps and async path queries.',
        events: [
            'pathfinding/find', 'pathfinding/find:async',
            'pathfinding/weight:set', 'pathfinding/cache:clear',
        ],
    };
    /** Sentinel cost value used to flag completely impassable cells. */
    static IMPASSABLE = IMPASSABLE;
    // ---------------------------------------------------------------------------
    // Internal state
    // ---------------------------------------------------------------------------
    _core = null;
    _options;
    /**
     * Cost grid built from the most recent `physics/tilemap:set` data.
     * `_grid[row][col]` is the movement cost for that cell (`IMPASSABLE` = blocked).
     */
    _grid = [];
    /** Number of rows in the current grid. */
    _rows = 0;
    /** Number of columns in the current grid. */
    _cols = 0;
    /** Tile size in pixels from the current tilemap (square tiles). */
    _tileSize = 0;
    /** Per-tile-value cost overrides applied on top of the shape-based defaults. */
    _weightOverrides = new Map();
    /** Per-tile-value raw data from the last tilemap (needed to re-apply weight overrides). */
    _tileValues = [];
    _tileShapes = {};
    /** Cached A* results keyed by `"fromRow,fromCol→toRow,toCol"`. */
    _cache = new Map();
    /** Worker bridge for async pathfinding.  `null` when no `workerUrl` was supplied. */
    _bridge = null;
    /**
     * `true` once the Worker has been initialised with the current cost grid.
     * While `false`, `pathfinding/find:async` falls back to the synchronous A*.
     */
    _workerReady = false;
    constructor(options = {}) {
        this._options = {
            directions: options.directions ?? 8,
            workerUrl: options.workerUrl ?? '',
        };
    }
    // ---------------------------------------------------------------------------
    // EnginePlugin lifecycle
    // ---------------------------------------------------------------------------
    init(core) {
        this._core = core;
        const { events } = core;
        // Spin up the Worker bridge if a URL was provided.
        if (this._options.workerUrl) {
            this._bridge = new WorkerBridge(this._options.workerUrl);
        }
        // Rebuild cost grid whenever the tile collision map changes.
        events.on(this.namespace, 'physics/tilemap:set', (params) => {
            this._buildGrid(params);
            this._cache.clear();
            // Sync the new grid to the Worker (fire-and-forget; _workerReady gates async calls).
            this._syncGridToWorker();
        });
        // O(1) update when a single tile is changed at runtime (e.g. door opening).
        // For collider layers this fires before the subsequent `physics/tilemap:set`
        // (which rebuilds the full grid); for non-collider layers with weight overrides
        // this is the only update that fires.
        events.on(this.namespace, 'tilemap/set-tile', (params) => {
            if (this._tileSize === 0)
                return;
            if (!this._inBounds(params.row, params.col))
                return;
            // Update the cached raw value and recompute this cell's cost.
            if (this._tileValues[params.row] !== undefined) {
                this._tileValues[params.row][params.col] = params.tileId;
            }
            const newCost = this._cellCost(params.tileId, this._tileShapes);
            if (this._grid[params.row] !== undefined) {
                this._grid[params.row][params.col] = newCost;
            }
            this._cache.clear();
            // Inform the Worker of the single-cell update (fire-and-forget).
            if (this._bridge && this._workerReady) {
                this._bridge
                    .run('tile:update', { row: params.row, col: params.col, cost: newCost })
                    .catch((err) => {
                    console.warn('[PathfindingManager] Worker tile:update failed:', err);
                });
            }
        });
        events.on(this.namespace, 'pathfinding/find', (params, output) => {
            const result = this._find(params);
            output.found = result.found;
            output.path = result.path;
            output.cost = result.cost;
            if (result.nearest !== undefined)
                output.nearest = result.nearest;
        });
        events.on(this.namespace, 'pathfinding/weight:set', (params) => {
            this._weightOverrides.set(params.tileId, params.cost);
            // Rebuild the grid so the new weight takes effect immediately.
            if (this._tileSize > 0) {
                this._rebuildGridFromCached();
                this._cache.clear();
                // Full grid changed — re-sync to Worker.
                this._syncGridToWorker();
            }
        });
        events.on(this.namespace, 'pathfinding/cache:clear', () => {
            this._cache.clear();
            // Also clear the Worker's internal cache.
            if (this._bridge && this._workerReady) {
                this._bridge.run('cache:clear', {}).catch((err) => {
                    console.warn('[PathfindingManager] Worker cache:clear failed:', err);
                });
            }
        });
        // Async variant: offloads to Worker when available, otherwise falls back
        // to the synchronous main-thread A* implementation.
        events.on(this.namespace, 'pathfinding/find:async', async (params, output) => {
            let result;
            if (this._bridge && this._workerReady) {
                // Collect dynamic obstacles on the main thread (entity/query requires it).
                const dynamicObstacleCells = params.includeDynamicObstacles
                    ? this._buildDynamicObstacleCells(params.tagFilter)
                    : [];
                result = await this._bridge.run('find', {
                    params: {
                        from: params.from,
                        to: params.to,
                        fallbackToNearest: params.fallbackToNearest,
                        smoothPath: params.smoothPath,
                        maxIterations: params.maxIterations,
                    },
                    dynamicObstacleCells,
                });
            }
            else {
                // Fallback: run synchronously on the main thread.
                result = this._find(params);
            }
            output.found = result.found;
            output.path = result.path;
            output.cost = result.cost;
            if (result.nearest !== undefined)
                output.nearest = result.nearest;
        });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
        this._grid = [];
        this._tileValues = [];
        this._tileShapes = {};
        this._cache.clear();
        this._weightOverrides.clear();
        this._rows = 0;
        this._cols = 0;
        this._tileSize = 0;
        this._core = null;
        // Shut down the Worker.
        this._bridge?.terminate();
        this._bridge = null;
        this._workerReady = false;
    }
    // ---------------------------------------------------------------------------
    // Public direct API
    // ---------------------------------------------------------------------------
    /**
     * Returns the movement cost grid cell at `(row, col)`, or `undefined` if
     * the coordinates are out of bounds.
     */
    getCellCost(row, col) {
        return this._grid[row]?.[col];
    }
    // ---------------------------------------------------------------------------
    // Private: Worker synchronisation
    // ---------------------------------------------------------------------------
    /**
     * Serialise the current cost grid as a `Float32Array` and send it to the
     * Worker via the `init` message type.  The method is fire-and-forget; once
     * the Worker acknowledges it sets `_workerReady = true`.
     *
     * A `Float32Array` is used so that values `> Float32.MAX` (i.e. `Infinity`)
     * round-trip correctly through `postMessage` — `Infinity` is preserved by
     * the structured-clone algorithm.
     */
    _syncGridToWorker() {
        if (!this._bridge || this._rows === 0 || this._cols === 0)
            return;
        this._workerReady = false;
        // Build a flat row-major Float32Array from the 2-D grid.
        const flat = new Float32Array(this._rows * this._cols);
        for (let r = 0; r < this._rows; r++) {
            for (let c = 0; c < this._cols; c++) {
                flat[r * this._cols + c] = this._grid[r][c];
            }
        }
        // Transfer the underlying ArrayBuffer to the Worker (zero-copy).
        this._bridge
            .run('init', {
            grid: flat,
            rows: this._rows,
            cols: this._cols,
            tileSize: this._tileSize,
            directions: this._options.directions,
        }, [flat.buffer])
            .then(() => {
            this._workerReady = true;
        })
            .catch((err) => {
            // Keep _workerReady = false so find:async falls back to sync A*.
            console.warn('[PathfindingManager] Worker initialisation failed:', err);
        });
    }
    /**
     * Collect entity tile positions for the Worker's dynamic-obstacle set.
     *
     * Returns an array of `[row, col]` pairs.  This must run on the main thread
     * because `entity/query` requires EventBus access.
     */
    _buildDynamicObstacleCells(tagFilter) {
        const cells = [];
        if (!this._core || this._tileSize === 0)
            return cells;
        const queryParams = tagFilter && tagFilter.length > 0 ? { tags: tagFilter } : {};
        const { output } = this._core.events.emitSync('entity/query', queryParams);
        const ts = this._tileSize;
        for (const entity of (output.entities ?? [])) {
            const row = Math.floor(entity.position.y / ts);
            const col = Math.floor(entity.position.x / ts);
            if (this._inBounds(row, col)) {
                cells.push([row, col]);
            }
        }
        return cells;
    }
    // ---------------------------------------------------------------------------
    // Private: grid construction
    // ---------------------------------------------------------------------------
    _buildGrid(params) {
        const { tileSize, layers, tileShapes } = params;
        this._tileSize = tileSize;
        this._tileShapes = tileShapes;
        // Store raw tile values for future weight-override rebuilds.
        this._tileValues = layers.map(row => [...row]);
        this._rows = layers.length;
        this._cols = layers.reduce((max, row) => Math.max(max, row.length), 0);
        this._grid = Array.from({ length: this._rows }, (_, r) => Array.from({ length: this._cols }, (_, c) => this._cellCost(layers[r]?.[c] ?? 0, tileShapes)));
    }
    /** Rebuild the grid from the cached raw tile values (after a weight override). */
    _rebuildGridFromCached() {
        this._grid = Array.from({ length: this._rows }, (_, r) => Array.from({ length: this._cols }, (_, c) => this._cellCost(this._tileValues[r]?.[c] ?? 0, this._tileShapes)));
    }
    /**
     * Compute the movement cost for a single cell.
     *
     * Priority (highest first):
     * 1. Explicit weight override for this tile value (`pathfinding/weight:set`).
     * 2. If the tile has a non-empty collision shape → `IMPASSABLE`.
     * 3. Default passable cost: `1`.
     */
    _cellCost(tileValue, tileShapes) {
        const override = this._weightOverrides.get(tileValue);
        if (override !== undefined)
            return override;
        const shape = tileShapes[tileValue];
        if (shape !== undefined && shape !== 'empty')
            return IMPASSABLE;
        return 1;
    }
    // ---------------------------------------------------------------------------
    // Private: A* search
    // ---------------------------------------------------------------------------
    _find(params) {
        const empty = { found: false, path: [], cost: 0 };
        if (this._tileSize === 0)
            return empty; // no tilemap loaded
        const ts = this._tileSize;
        const fromRow = Math.floor(params.from.y / ts);
        const fromCol = Math.floor(params.from.x / ts);
        const toRow = Math.floor(params.to.y / ts);
        const toCol = Math.floor(params.to.x / ts);
        // Already at destination
        if (fromRow === toRow && fromCol === toCol) {
            const pt = this._tileCenter(toRow, toCol);
            return { found: true, path: [pt], cost: 0 };
        }
        // Bounds check
        if (!this._inBounds(fromRow, fromCol) || !this._inBounds(toRow, toCol))
            return empty;
        // Start-cell impassability check — e.g. entity was knocked into a wall.
        if (this._grid[fromRow][fromCol] >= IMPASSABLE)
            return empty;
        // Goal cell: try fallback to nearest passable when requested.
        let effectiveToRow = toRow;
        let effectiveToCol = toCol;
        let nearest;
        if (this._grid[toRow][toCol] >= IMPASSABLE) {
            if (!params.fallbackToNearest)
                return empty;
            const fallback = this._findNearestPassable(toRow, toCol);
            if (!fallback)
                return empty;
            effectiveToRow = fallback.row;
            effectiveToCol = fallback.col;
            nearest = this._tileCenter(effectiveToRow, effectiveToCol);
        }
        // Cache lookup (static searches only)
        const cacheKey = `${fromRow},${fromCol}→${effectiveToRow},${effectiveToCol}`;
        if (!params.includeDynamicObstacles) {
            const cached = this._cacheGet(cacheKey);
            if (cached) {
                // Re-attach nearest when serving from cache.
                return nearest !== undefined ? { ...cached, nearest } : cached;
            }
        }
        // Build dynamic obstacle set if requested
        const dynamicBlocked = params.includeDynamicObstacles
            ? this._buildDynamicObstacles(params.tagFilter)
            : null;
        const maxIter = params.maxIterations ?? 10_000;
        let result = this._astar(fromRow, fromCol, effectiveToRow, effectiveToCol, dynamicBlocked, maxIter);
        // Optional path smoothing (string-pulling)
        if (result.found && params.smoothPath && result.path.length > 2) {
            result = { ...result, path: this._smoothPath(result.path) };
        }
        // Attach nearest if a fallback cell was used
        if (nearest !== undefined && result.found) {
            result = { ...result, nearest };
        }
        // Cache static results (store without nearest so cache is goal-agnostic)
        if (!params.includeDynamicObstacles) {
            const toCache = nearest !== undefined ? { found: result.found, path: result.path, cost: result.cost } : result;
            this._cacheSet(cacheKey, toCache);
        }
        return result;
    }
    _astar(fromRow, fromCol, toRow, toCol, dynamicBlocked, maxIter) {
        const empty = { found: false, path: [], cost: 0 };
        const openSet = new MinHeap();
        /** Cheapest g-cost found for each cell so far; key = `"row,col"`. */
        const gScores = new Map();
        /** Cells already processed. */
        const closed = new Set();
        const startNode = {
            row: fromRow, col: fromCol,
            g: 0,
            f: this._heuristic(fromRow, fromCol, toRow, toCol),
            parent: null,
        };
        const startKey = `${fromRow},${fromCol}`;
        openSet.push(startNode);
        gScores.set(startKey, 0);
        let iterations = 0;
        while (openSet.size > 0) {
            if (++iterations > maxIter)
                return empty;
            const current = openSet.pop();
            const currentKey = `${current.row},${current.col}`;
            if (closed.has(currentKey))
                continue;
            closed.add(currentKey);
            if (current.row === toRow && current.col === toCol) {
                // Reconstruct path
                return this._reconstructPath(current);
            }
            for (const [nr, nc, moveCost] of this._neighbours(current.row, current.col)) {
                const nKey = `${nr},${nc}`;
                if (closed.has(nKey))
                    continue;
                // Check static impassability
                const cellCost = this._grid[nr][nc];
                if (cellCost >= IMPASSABLE)
                    continue;
                // Check dynamic obstacles (skip goal cell — we want to reach it)
                if (dynamicBlocked?.has(nKey) && !(nr === toRow && nc === toCol))
                    continue;
                const gNew = current.g + moveCost * cellCost;
                const prevG = gScores.get(nKey) ?? Infinity;
                if (gNew < prevG) {
                    gScores.set(nKey, gNew);
                    openSet.push({
                        row: nr, col: nc,
                        g: gNew,
                        f: gNew + this._heuristic(nr, nc, toRow, toCol),
                        parent: current,
                    });
                }
            }
        }
        return empty; // no path found
    }
    _reconstructPath(node) {
        const cells = [];
        let cur = node;
        while (cur !== null) {
            cells.push(cur);
            cur = cur.parent;
        }
        cells.reverse();
        const path = cells.map(n => this._tileCenter(n.row, n.col));
        return { found: true, path, cost: node.g };
    }
    /** Chebyshev distance (8-dir) or Manhattan distance (4-dir). */
    _heuristic(r1, c1, r2, c2) {
        const dr = Math.abs(r1 - r2);
        const dc = Math.abs(c1 - c2);
        if (this._options.directions === 8) {
            return Math.max(dr, dc); // Chebyshev
        }
        return dr + dc; // Manhattan
    }
    /**
     * Returns the neighbouring cells of `(row, col)` that are within the grid
     * bounds, together with the base movement cost (1 for cardinal, √2 ≈ 1.414
     * for diagonal — the actual per-cell cost is multiplied in `_astar`).
     */
    _neighbours(row, col) {
        const dirs = this._options.directions === 8
            ? DIRS_8
            : DIRS_4;
        const result = [];
        for (const [dr, dc, cost] of dirs) {
            const nr = row + dr;
            const nc = col + dc;
            if (this._inBounds(nr, nc)) {
                result.push([nr, nc, cost]);
            }
        }
        return result;
    }
    _inBounds(row, col) {
        return row >= 0 && row < this._rows && col >= 0 && col < this._cols;
    }
    /** World-pixel coordinate of the centre of tile `(row, col)`. */
    _tileCenter(row, col) {
        const half = this._tileSize / 2;
        return {
            x: col * this._tileSize + half,
            y: row * this._tileSize + half,
        };
    }
    // ---------------------------------------------------------------------------
    // Private: dynamic obstacles
    // ---------------------------------------------------------------------------
    /**
     * Query active entities and return a set of tile-cell keys (`"row,col"`)
     * that are occupied by those entities.
     *
     * Pass `tagFilter` to restrict which entities are considered.  Without a
     * filter every active entity is included; use tags such as `'obstacle'` to
     * exclude decorative sprites and HUD anchors.
     *
     * The method queries entity positions via `entity/query` and uses its own
     * simple AABB-to-tile conversion without depending on `CollisionManager`
     * internals.
     */
    _buildDynamicObstacles(tagFilter) {
        const blocked = new Set();
        if (!this._core || this._tileSize === 0)
            return blocked;
        const queryParams = tagFilter && tagFilter.length > 0
            ? { tags: tagFilter }
            : {};
        const { output } = this._core.events.emitSync('entity/query', queryParams);
        const ts = this._tileSize;
        for (const entity of (output.entities ?? [])) {
            const row = Math.floor(entity.position.y / ts);
            const col = Math.floor(entity.position.x / ts);
            if (this._inBounds(row, col)) {
                blocked.add(`${row},${col}`);
            }
        }
        return blocked;
    }
    // ---------------------------------------------------------------------------
    // Private: LRU cache helpers
    // ---------------------------------------------------------------------------
    /**
     * Retrieve a cached path, refreshing its recency so it is not evicted before
     * less-recently-used entries.
     */
    _cacheGet(key) {
        const value = this._cache.get(key);
        if (value === undefined)
            return undefined;
        // Move to most-recently-used position by re-inserting.
        this._cache.delete(key);
        this._cache.set(key, value);
        return value;
    }
    /**
     * Insert a path into the LRU cache.  When the cache exceeds
     * {@link CACHE_MAX_SIZE} the oldest (least-recently-used) entry is evicted.
     */
    _cacheSet(key, value) {
        if (this._cache.has(key))
            this._cache.delete(key);
        this._cache.set(key, value);
        if (this._cache.size > CACHE_MAX_SIZE) {
            // Map preserves insertion order; the first key is the oldest.
            this._cache.delete(this._cache.keys().next().value);
        }
    }
    // ---------------------------------------------------------------------------
    // Private: fallback nearest-passable cell (BFS)
    // ---------------------------------------------------------------------------
    /**
     * BFS outward from `(row, col)` to find the closest cell in the static grid
     * that is passable (cost < {@link IMPASSABLE}).
     *
     * Returns `null` when no passable cell is found within
     * {@link NEAREST_MAX_EXPANSIONS} steps.
     */
    _findNearestPassable(row, col) {
        const visited = new Set();
        const queue = [{ row, col }];
        visited.add(`${row},${col}`);
        let expansions = 0;
        while (queue.length > 0 && expansions < NEAREST_MAX_EXPANSIONS) {
            const current = queue.shift();
            expansions++;
            if ((this._grid[current.row]?.[current.col] ?? IMPASSABLE) < IMPASSABLE) {
                return current;
            }
            for (const [dr, dc] of DIRS_4) {
                const nr = current.row + dr;
                const nc = current.col + dc;
                const nKey = `${nr},${nc}`;
                if (this._inBounds(nr, nc) && !visited.has(nKey)) {
                    visited.add(nKey);
                    queue.push({ row: nr, col: nc });
                }
            }
        }
        return null;
    }
    // ---------------------------------------------------------------------------
    // Private: path smoothing (string-pulling)
    // ---------------------------------------------------------------------------
    /**
     * Remove redundant intermediate waypoints using a string-pulling pass.
     *
     * Two consecutive waypoints are merged when there is unobstructed
     * line-of-sight between them on the tile grid (verified with Bresenham's
     * line algorithm).  This eliminates the staircase artifacts typical of
     * diagonal A* paths.
     */
    _smoothPath(path) {
        if (path.length <= 2)
            return path;
        const result = [path[0]];
        let anchor = 0;
        for (let i = 2; i < path.length; i++) {
            if (!this._hasLoS(path[anchor], path[i])) {
                // No LoS from anchor to i — commit the previous waypoint.
                result.push(path[i - 1]);
                anchor = i - 1;
            }
        }
        result.push(path[path.length - 1]);
        return result;
    }
    /**
     * Returns `true` when every tile cell along the straight line from `a` to
     * `b` (in world-pixel coordinates) is passable in the static grid.
     *
     * Uses Bresenham's line algorithm for integer tile-grid traversal.
     */
    _hasLoS(a, b) {
        const ts = this._tileSize;
        let r0 = Math.floor(a.y / ts);
        let c0 = Math.floor(a.x / ts);
        const r1 = Math.floor(b.y / ts);
        const c1 = Math.floor(b.x / ts);
        const dr = Math.abs(r1 - r0);
        const dc = Math.abs(c1 - c0);
        const sr = r0 < r1 ? 1 : -1;
        const sc = c0 < c1 ? 1 : -1;
        let err = dr - dc;
        while (true) {
            if ((this._grid[r0]?.[c0] ?? IMPASSABLE) >= IMPASSABLE)
                return false;
            if (r0 === r1 && c0 === c1)
                break;
            const e2 = 2 * err;
            if (e2 > -dc) {
                err -= dc;
                r0 += sr;
            }
            if (e2 < dr) {
                err += dr;
                c0 += sc;
            }
        }
        return true;
    }
}
// ---------------------------------------------------------------------------
// Direction tables
// ---------------------------------------------------------------------------
/** [dRow, dCol, movementCost] for 4-directional movement. */
const DIRS_4 = [
    [-1, 0, 1], // up
    [1, 0, 1], // down
    [0, -1, 1], // left
    [0, 1, 1], // right
];
/** [dRow, dCol, movementCost] for 8-directional movement. */
const DIRS_8 = [
    [-1, 0, 1], // up
    [1, 0, 1], // down
    [0, -1, 1], // left
    [0, 1, 1], // right
    [-1, -1, Math.SQRT2], // up-left
    [-1, 1, Math.SQRT2], // up-right
    [1, -1, Math.SQRT2], // down-left
    [1, 1, Math.SQRT2], // down-right
];
