// @inkshot/engine/dist/plugins/gameplay/CutscenePlugin.js
// ---------------------------------------------------------------------------
// CutscenePlugin
// ---------------------------------------------------------------------------
/**
 * Plugin that drives **scripted cutscenes** using a step-based timeline.
 *
 * A cutscene is a named sequence of {@link CutsceneStep} objects.  Each step
 * is executed in order; some steps complete instantly while others wait for
 * a timer (`wait`), a script (`script`), or a camera tween
 * (`camera-move` / `camera-zoom`).
 *
 * Cutscenes can optionally be **skipped** by the player (set
 * `CutsceneDef.skippable = true` and emit `cutscene/skip`).
 *
 * The `parallel` step kind lets you run multiple sub-sequences
 * simultaneously and wait for all of them to finish.
 *
 * ### Quick start
 * ```ts
 * import { createEngine, CutscenePlugin } from 'inkshot-engine';
 *
 * const cutscene = new CutscenePlugin();
 * const { core } = await createEngine({ plugins: [cutscene] });
 *
 * core.events.emitSync('cutscene/define', {
 *   cutscene: {
 *     id: 'intro',
 *     skippable: true,
 *     steps: [
 *       { kind: 'lock-input' },
 *       { kind: 'camera-move', x: 500, y: 300, duration: 1000 },
 *       { kind: 'wait', duration: 2000 },
 *       { kind: 'camera-follow', entityId: 'player' },
 *       { kind: 'unlock-input' },
 *     ],
 *   },
 * });
 *
 * core.events.emitSync('cutscene/play', { id: 'intro' });
 * ```
 *
 * ### EventBus API
 *
 * | Event                    | Params / Output                                         |
 * |--------------------------|---------------------------------------------------------|
 * | `cutscene/define`        | `CutsceneDefineParams`                                  |
 * | `cutscene/play`          | `CutscenePlayParams`                                    |
 * | `cutscene/skip`          | `{}`                                                    |
 * | `cutscene/stop`          | `{}`                                                    |
 * | `cutscene/state`         | `{} → CutsceneStateOutput`                              |
 * | `cutscene/started`       | `CutsceneStartedParams` (notification)                  |
 * | `cutscene/ended`         | `CutsceneEndedParams` (notification)                    |
 * | `cutscene/step:started`  | `CutsceneStepStartedParams` (notification)              |
 * | `cutscene/step:ended`    | `CutsceneStepEndedParams` (notification)                |
 */
export class CutscenePlugin {
    namespace = 'cutscene';
    editorMeta = {
        displayName: 'Cutscene Plugin',
        icon: 'cutscene',
        description: 'Plays scripted cutscene sequences with camera moves, waits, and event steps.',
        events: [
            'cutscene/define', 'cutscene/play', 'cutscene/stop', 'cutscene/skip', 'cutscene/state',
        ],
        schemas: {
            cutscene: {
                folder: 'cutscenes',
                displayName: 'Cutscene Definition',
            },
        },
    };
    _definitions = new Map();
    _active = null;
    // ---------------------------------------------------------------------------
    // EnginePlugin lifecycle
    // ---------------------------------------------------------------------------
    init(core) {
        // ── Event handlers ────────────────────────────────────────────────────
        core.events.on('cutscene', 'cutscene/define', (p) => {
            this._definitions.set(p.cutscene.id, p.cutscene);
        });
        core.events.on('cutscene', 'cutscene/play', (p) => {
            if (this._active) {
                console.warn(`[CutscenePlugin] A cutscene ("${this._active.def.id}") is already playing. ` +
                    `Call cutscene/stop first.`);
                return;
            }
            const def = this._definitions.get(p.id);
            if (!def) {
                console.warn(`[CutscenePlugin] Unknown cutscene: "${p.id}".`);
                return;
            }
            this._startCutscene(def, core);
        });
        core.events.on('cutscene', 'cutscene/skip', () => {
            if (this._active && this._active.def.skippable !== false) {
                this._active.skip = true;
            }
        });
        core.events.on('cutscene', 'cutscene/stop', () => {
            if (this._active) {
                this._endCutscene(core, true);
            }
        });
        core.events.on('cutscene', 'cutscene/state', (_p, output) => {
            const state = this._getState();
            output.status = state.status;
            output.cutsceneId = state.cutsceneId;
            output.stepIndex = state.stepIndex;
        });
        // ── Advance active cutscene every fixed tick ───────────────────────────
        core.events.on('cutscene', 'core/update', (p) => {
            if (this._active) {
                this._advance(this._active, p.dt, core);
            }
        });
    }
    destroy(core) {
        this._active = null;
        core.events.removeNamespace('cutscene');
    }
    // ---------------------------------------------------------------------------
    // Direct accessors (pull API)
    // ---------------------------------------------------------------------------
    getState() {
        return this._getState();
    }
    isPlaying() {
        return this._active !== null;
    }
    // ---------------------------------------------------------------------------
    // Private — cutscene lifecycle
    // ---------------------------------------------------------------------------
    _startCutscene(def, core) {
        this._active = {
            def,
            stepIndex: -1,
            elapsed: 0,
            waitingForScript: false,
            waitingForTween: false,
            stepRemaining: 0,
            skip: false,
            parallelRunners: null,
        };
        core.events.emitSync('cutscene/started', { id: def.id });
        this._nextStep(this._active, core);
    }
    _endCutscene(core, skipped) {
        if (!this._active)
            return;
        const id = this._active.def.id;
        this._active = null;
        core.events.emitSync('cutscene/ended', { id, skipped });
    }
    _nextStep(active, core) {
        active.stepIndex += 1;
        active.elapsed = 0;
        active.waitingForScript = false;
        active.waitingForTween = false;
        active.stepRemaining = 0;
        active.parallelRunners = null;
        const { def, stepIndex } = active;
        // All steps exhausted → cutscene done.
        if (stepIndex >= def.steps.length) {
            this._endCutscene(core, false);
            return;
        }
        const step = def.steps[stepIndex];
        core.events.emitSync('cutscene/step:started', {
            cutsceneId: def.id,
            stepIndex,
            step,
        });
        this._executeStep(active, step, core);
    }
    _executeStep(active, step, core) {
        switch (step.kind) {
            case 'wait':
                active.stepRemaining = step.duration;
                return;
            case 'camera-move': {
                const dur = step.duration ?? 500;
                if (dur <= 0) {
                    core.events.emitSync('camera/move', { x: step.x, y: step.y });
                    this._completeStep(active, core);
                }
                else {
                    // Start a camera tween via TweenManager (best-effort; move immediately
                    // if TweenManager is absent).
                    core.events.emitSync('tween/to', {
                        target: '_camera_',
                        props: { x: step.x, y: step.y },
                        duration: dur,
                        onUpdate: (vals) => {
                            core.events.emitSync('camera/move', { x: vals.x, y: vals.y });
                        },
                        id: `cutscene_cam_move_${active.def.id}`,
                    });
                    active.stepRemaining = dur;
                    active.waitingForTween = true;
                }
                return;
            }
            case 'camera-zoom': {
                const dur = step.duration ?? 500;
                if (dur <= 0) {
                    core.events.emitSync('camera/zoom', { zoom: step.zoom });
                    this._completeStep(active, core);
                }
                else {
                    active.stepRemaining = dur;
                    active.waitingForTween = true;
                }
                return;
            }
            case 'camera-shake':
                core.events.emitSync('camera/shake', {
                    intensity: step.intensity,
                    duration: step.duration,
                    decay: step.decay,
                });
                this._completeStep(active, core);
                return;
            case 'camera-follow': {
                if (step.entityId === null) {
                    core.events.emitSync('camera/follow', { target: null });
                }
                else {
                    // Retrieve the entity's position object and pass it as a follow target.
                    const { output: entityOut } = core.events.emitSync('entity/query', { tags: [] });
                    const found = (entityOut.entities ?? []).find((e) => e.id === step.entityId);
                    core.events.emitSync('camera/follow', { target: found?.position ?? null });
                }
                this._completeStep(active, core);
                return;
            }
            case 'emit':
                core.events.emitSync(step.event, step.params ?? {});
                this._completeStep(active, core);
                return;
            case 'script': {
                const scriptOut = {};
                core.events.emitSync('script/run', { scriptId: step.scriptId, context: step.context ?? {} }, scriptOut);
                active.waitingForScript = true;
                // Listen for the script to end, then advance.
                const instanceId = scriptOut.instanceId;
                if (instanceId) {
                    core.events.once('cutscene', 'script/ended', (p) => {
                        if (p.instanceId === instanceId && this._active === active) {
                            active.waitingForScript = false;
                            this._completeStep(active, core);
                        }
                    });
                }
                else {
                    // Script system absent — advance immediately.
                    active.waitingForScript = false;
                    this._completeStep(active, core);
                }
                return;
            }
            case 'lock-input':
                core.events.emitSync('input/lock', {});
                this._completeStep(active, core);
                return;
            case 'unlock-input':
                core.events.emitSync('input/unlock', {});
                this._completeStep(active, core);
                return;
            case 'parallel': {
                active.parallelRunners = step.steps.map((s) => ({
                    steps: [s],
                    index: -1,
                    elapsed: 0,
                    stepRemaining: 0,
                    waitingForScript: false,
                    done: false,
                }));
                for (const runner of active.parallelRunners) {
                    this._advanceRunner(runner, 0, active, core);
                }
                return;
            }
        }
    }
    _completeStep(active, core) {
        core.events.emitSync('cutscene/step:ended', {
            cutsceneId: active.def.id,
            stepIndex: active.stepIndex,
        });
        this._nextStep(active, core);
    }
    // ---------------------------------------------------------------------------
    // Private — update loop
    // ---------------------------------------------------------------------------
    _advance(active, dt, core) {
        // Skip requested — jump to end.
        if (active.skip) {
            this._endCutscene(core, true);
            return;
        }
        // ── Parallel runner ──────────────────────────────────────────────────
        if (active.parallelRunners) {
            let allDone = true;
            for (const runner of active.parallelRunners) {
                if (!runner.done) {
                    this._advanceRunner(runner, dt, active, core);
                    if (!runner.done)
                        allDone = false;
                }
            }
            if (allDone) {
                active.parallelRunners = null;
                this._completeStep(active, core);
            }
            return;
        }
        // ── Waiting for script ───────────────────────────────────────────────
        if (active.waitingForScript)
            return;
        // ── Timed step (wait / camera tween) ────────────────────────────────
        if (active.stepRemaining > 0) {
            active.stepRemaining -= dt;
            if (active.stepRemaining <= 0) {
                active.stepRemaining = 0;
                this._completeStep(active, core);
            }
        }
    }
    _advanceRunner(runner, dt, active, core) {
        if (runner.done)
            return;
        if (runner.stepRemaining > 0) {
            runner.stepRemaining -= dt;
            if (runner.stepRemaining <= 0) {
                runner.stepRemaining = 0;
                runner.done = true;
            }
            return;
        }
        if (runner.waitingForScript)
            return;
        runner.index += 1;
        if (runner.index >= runner.steps.length) {
            runner.done = true;
            return;
        }
        const step = runner.steps[runner.index];
        // For parallel sub-steps, only 'wait' and instant steps are supported.
        if (step.kind === 'wait') {
            runner.stepRemaining = step.duration;
        }
        else {
            // Instant step — execute and immediately mark done.
            this._executeStep(active, step, core);
            runner.done = true;
        }
    }
    // ---------------------------------------------------------------------------
    // Private — helpers
    // ---------------------------------------------------------------------------
    _getState() {
        if (!this._active) {
            return { status: 'idle', cutsceneId: null, stepIndex: null };
        }
        const status = this._active.skip ? 'skipping' : 'playing';
        return {
            status,
            cutsceneId: this._active.def.id,
            stepIndex: this._active.stepIndex,
        };
    }
}
