// @inkshot/engine/dist/index.js
// Settings system
export { SettingsManager } from './plugins/data/SettingsManager.js';
// Audio system
export { AudioManager } from './plugins/audio/AudioManager.js';
// Asset system
export { ResourceManager } from './plugins/data/ResourceManager.js';
// Data system
export { DataManager } from './plugins/data/DataManager.js';
// Localisation system
export { LocalizationManager } from './plugins/data/LocalizationManager.js';
// Core
export { Core } from './core/Core.js';
// Worker bridge (generic Web Worker offloading utility)
export { WorkerBridge } from './core/WorkerBridge.js';
// Input system
export { InputManager } from './plugins/input/InputManager.js';
export { InputRecorder } from './plugins/input/InputRecorder.js';
// Game state system
export { GameStateManager } from './plugins/gameplay/GameStateManager.js';
// Event Bus
export { EventBus } from './core/EventBus.js';
// Save system
export { SaveManager } from './plugins/save/SaveManager.js';
export { LocalStorageSaveAdapter } from './plugins/save/LocalStorageSaveAdapter.js';
export { IndexedDBSaveAdapter } from './plugins/save/IndexedDBSaveAdapter.js';
export { SaveMigrationPlugin } from './plugins/save/SaveMigrationPlugin.js';
// Scene system
export { SceneManager } from './plugins/scene/SceneManager.js';
export { LoadingScreen } from './plugins/ui/LoadingScreen.js';
// Physics system (unified adapter interface + kinematic built-in backend)
export { KinematicPhysicsAdapter } from './plugins/physics/KinematicPhysicsAdapter.js';
export { MatterPhysicsAdapter } from './plugins/physics/MatterPhysicsAdapter.js';
export { RapierPhysicsAdapter } from './plugins/physics/RapierPhysicsAdapter.js';
export { CollisionLayer } from './types/physics.js';
// Entity system
export { EntityManager } from './plugins/entity/EntityManager.js';
export { SpriteAnimator } from './plugins/entity/SpriteAnimator.js';
// Engine factory (primary public entry point)
export { createEngine, sortPluginsByDependency } from './createEngine.js';
// Tilemap system
export { TilemapManager } from './plugins/tilemap/TilemapManager.js';
export { AutotileBit } from './types/tilemap.js';
// Tilemap editor plugin
export { TilemapEditorPlugin } from './plugins/tilemap/TilemapEditorPlugin.js';
export { TilemapEditorOverlayPlugin } from './plugins/tilemap/TilemapEditorOverlayPlugin.js';
// Scene editor plugin
export { SceneEditorPlugin } from './plugins/scene/SceneEditorPlugin.js';
// Script node editor plugin
export { ScriptNodeEditorPlugin } from './plugins/rpg/ScriptNodeEditorPlugin.js';
// Tween system
export { TweenManager, Tween, Easing } from './plugins/animation/TweenManager.js';
export { Timeline } from './plugins/animation/Timeline.js';
// Particle system
export { ParticleManager } from './plugins/world/ParticleManager.js';
// Rendering
export { Renderer } from './rendering/Renderer.js';
export { RenderPipeline } from './rendering/RenderPipeline.js';
export { PostFxPipeline } from './rendering/PostFxPipeline.js';
export { ShaderPass } from './rendering/ShaderPass.js';
export { AnimationSystem } from './rendering/AnimationSystem.js';
export { Camera } from './rendering/Camera.js';
export { ObjectPool } from './rendering/ObjectPool.js';
export { LAYER_Z_INDEX } from './rendering/layers.js';
// Timer / Scheduler system
export { TimerManager } from './plugins/gameplay/TimerManager.js';
// Pathfinding system
export { PathfindingManager } from './plugins/world/PathfindingManager.js';
// UI system
export { UIManager } from './plugins/ui/UIManager.js';
// Actor system
export { ActorManager } from './plugins/rpg/ActorManager.js';
// Player controller
export { PlayerController } from './plugins/rpg/PlayerController.js';
// Variable store system
export { VariableStoreManager } from './plugins/rpg/VariableStoreManager.js';
// Script system
export { ScriptManager } from './plugins/rpg/ScriptManager.js';
// Dialogue system
export { DialogueManager } from './plugins/rpg/DialogueManager.js';
export { parseDialogueMarkup, buildTextSegments, getSpeedAtIndex, } from './plugins/rpg/DialogueMarkupParser.js';
// Debug / Dev Tools
export { DebugPlugin } from './plugins/debug/DebugPlugin.js';
// Lighting system
export { LightingPlugin } from './plugins/world/LightingPlugin.js';
// Parallax system
export { ParallaxPlugin } from './plugins/world/ParallaxPlugin.js';
// Tiled map loader
export { loadTiledMap } from './plugins/tilemap/TiledLoader.js';
// Cutscene system
export { CutscenePlugin } from './plugins/gameplay/CutscenePlugin.js';
// Minimap system
export { MinimapPlugin } from './plugins/ui/MinimapPlugin.js';
// Achievement system
export { AchievementPlugin } from './plugins/gameplay/AchievementPlugin.js';
// Fog of War system
export { FogOfWarPlugin } from './plugins/world/FogOfWarPlugin.js';
// ── RPG Systems ───────────────────────────────────────────────────────────────
export { StatsSystem } from './plugins/rpg/StatsSystem.js';
export { InventorySystem } from './plugins/rpg/InventorySystem.js';
export { ShopSystem } from './plugins/rpg/ShopSystem.js';
export { ExperienceSystem } from './plugins/rpg/ExperienceSystem.js';
export { BattleSystem } from './plugins/rpg/BattleSystem.js';
export { RpgMenuSystem } from './plugins/rpg/RpgMenuSystem.js';
// ── RPG factory + bundle ───────────────────────────────────────────────────────
export { buildRpgPluginBundle, RPG_PLUGIN_BUNDLE, createRpgEngine, } from './rpg/createRpgEngine.js';
// ── RPG data loader ────────────────────────────────────────────────────────────
export { loadRpgData, registerRpgData } from './rpg/data/index.js';
