// @inkshot/engine/dist/rpg/createRpgEngine.js
import { createEngine } from '../createEngine.js';
import { registerRpgData } from './data/index.js';
// Core plugins already in the engine
import { AudioManager } from '../plugins/audio/AudioManager.js';
import { ResourceManager } from '../plugins/data/ResourceManager.js';
import { DataManager } from '../plugins/data/DataManager.js';
import { InputManager } from '../plugins/input/InputManager.js';
import { GameStateManager } from '../plugins/gameplay/GameStateManager.js';
import { SceneManager } from '../plugins/scene/SceneManager.js';
import { EntityManager } from '../plugins/entity/EntityManager.js';
import { TilemapManager } from '../plugins/tilemap/TilemapManager.js';
import { TweenManager } from '../plugins/animation/TweenManager.js';
import { TimerManager } from '../plugins/gameplay/TimerManager.js';
import { UIManager } from '../plugins/ui/UIManager.js';
import { SaveManager } from '../plugins/save/SaveManager.js';
import { LocalStorageSaveAdapter } from '../plugins/save/LocalStorageSaveAdapter.js';
import { LocalizationManager } from '../plugins/data/LocalizationManager.js';
import { ParticleManager } from '../plugins/world/ParticleManager.js';
import { PathfindingManager } from '../plugins/world/PathfindingManager.js';
import { AchievementPlugin } from '../plugins/gameplay/AchievementPlugin.js';
import { LightingPlugin } from '../plugins/world/LightingPlugin.js';
import { FogOfWarPlugin } from '../plugins/world/FogOfWarPlugin.js';
import { MinimapPlugin } from '../plugins/ui/MinimapPlugin.js';
// RPG plugins
import { VariableStoreManager } from '../plugins/rpg/VariableStoreManager.js';
import { ScriptManager } from '../plugins/rpg/ScriptManager.js';
import { DialogueManager } from '../plugins/rpg/DialogueManager.js';
import { ActorManager } from '../plugins/rpg/ActorManager.js';
import { PlayerController } from '../plugins/rpg/PlayerController.js';
import { StatsSystem } from '../plugins/rpg/StatsSystem.js';
import { InventorySystem } from '../plugins/rpg/InventorySystem.js';
import { ShopSystem } from '../plugins/rpg/ShopSystem.js';
import { ExperienceSystem } from '../plugins/rpg/ExperienceSystem.js';
import { BattleSystem } from '../plugins/rpg/BattleSystem.js';
import { RpgMenuSystem } from '../plugins/rpg/RpgMenuSystem.js';
// ---------------------------------------------------------------------------
// RPG_PLUGIN_BUNDLE
// ---------------------------------------------------------------------------
/**
 * Build the ordered list of all built-in RPG plugins.
 *
 * Use {@link createRpgEngine} for a one-liner; call this directly only when
 * you need to compose the plugin list manually.
 */
export function buildRpgPluginBundle(opts = {}) {
    return [
        new ResourceManager(),
        new DataManager(),
        new AudioManager(),
        new InputManager(),
        new GameStateManager(),
        new LocalStorageSaveAdapter(),
        new SaveManager(),
        new LocalizationManager(),
        new SceneManager(),
        new EntityManager(),
        new TilemapManager(),
        new TweenManager(),
        new TimerManager(),
        new ParticleManager(),
        new PathfindingManager(),
        new UIManager(),
        new AchievementPlugin(),
        new LightingPlugin(),
        new FogOfWarPlugin(),
        new MinimapPlugin(),
        new VariableStoreManager(),
        new ScriptManager(),
        new DialogueManager(),
        new ActorManager(),
        new PlayerController(),
        new StatsSystem(),
        new InventorySystem(),
        new ShopSystem(opts.shop),
        new ExperienceSystem(),
        new BattleSystem(),
        new RpgMenuSystem(opts.rpgMenu),
    ];
}
/**
 * Ordered array of default RPG plugin instances (no custom options).
 *
 * > **Note**: This is a snapshot created at module load time.
 * > For custom per-system options call {@link buildRpgPluginBundle} or
 * > {@link createRpgEngine}.
 */
export const RPG_PLUGIN_BUNDLE = buildRpgPluginBundle();
// ---------------------------------------------------------------------------
// createRpgEngine
// ---------------------------------------------------------------------------
/**
 * **One-liner RPG engine startup.**
 *
 * Creates and initialises the Inkshot engine with the full RPG plugin bundle
 * (audio, input, save, scenes, tilemap, dialogue, scripts, actors, player,
 * stats, inventory, shop, experience, battle, and menu) in a single call.
 *
 * ```ts
 * import { createRpgEngine } from '@inkshot/engine/rpg';
 *
 * const { core, rpg } = await createRpgEngine({
 *   container: '#app',
 *   width: 1280,
 *   height: 720,
 *   dataRoot: '/assets/',
 * });
 *
 * core.events.emitSync('stats/base:set', { actorId: 'hero', patch: { hp: 100, hpMax: 100 } });
 * ```
 */
export async function createRpgEngine(options = {}) {
    const { rpgMenu, shop, gameData, extraPlugins = [], plugins: userPlugins = [], ...engineOptions } = options;
    const bundle = buildRpgPluginBundle({ rpgMenu, shop });
    const allPlugins = [...bundle, ...extraPlugins, ...userPlugins];
    const base = await createEngine({ ...engineOptions, plugins: allPlugins });
    const findPlugin = (ns) => {
        const p = base.plugins.find((pl) => pl.namespace === ns);
        if (!p)
            throw new Error(`[createRpgEngine] Plugin "${ns}" was not initialised.`);
        return p;
    };
    const instance = {
        ...base,
        rpg: {
            stats: findPlugin('stats'),
            inventory: findPlugin('inventory'),
            shop: findPlugin('shop'),
            exp: findPlugin('exp'),
            battle: findPlugin('battle'),
            menu: findPlugin('rpgmenu'),
            variableStore: findPlugin('store'),
            script: findPlugin('script'),
            dialogue: findPlugin('dialogue'),
            actor: findPlugin('actor'),
        },
    };
    if (gameData) {
        registerRpgData(base.core, gameData);
    }
    return instance;
}
