// @inkshot/engine/dist/plugins/scene/SceneEditorPlugin.js
// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------
const MAX_UNDO_HISTORY = 100;
/** Name assigned to the Tiled-compatible object layer produced by `sceneeditor/export`. */
const EXPORT_LAYER_NAME = 'objects';
// ---------------------------------------------------------------------------
// SceneEditorPlugin
// ---------------------------------------------------------------------------
/**
 * In-engine scene editor plugin.
 *
 * Provides a workflow for placing, moving, and removing scene objects driven
 * exclusively through the {@link EventBus}.
 *
 * ### EventBus API
 *
 * | Event                              | Params / Output                              |
 * |------------------------------------|----------------------------------------------|
 * | `sceneeditor/open`                 | `SceneEditorOpenParams`                      |
 * | `sceneeditor/close`                | `SceneEditorCloseParams`                     |
 * | `sceneeditor/tool:set`             | `SceneEditorToolSetParams`                   |
 * | `sceneeditor/actor-type:select`    | `SceneEditorActorTypeSelectParams`           |
 * | `sceneeditor/object:place`         | `SceneEditorObjectPlaceParams → Output`      |
 * | `sceneeditor/object:select`        | `SceneEditorObjectSelectParams`              |
 * | `sceneeditor/object:move`          | `SceneEditorObjectMoveParams`                |
 * | `sceneeditor/object:remove`        | `SceneEditorObjectRemoveParams`              |
 * | `sceneeditor/undo`                 | `SceneEditorUndoParams`                      |
 * | `sceneeditor/redo`                 | `SceneEditorRedoParams`                      |
 * | `sceneeditor/export`               | `{} → SceneEditorExportOutput`               |
 * | `sceneeditor/state`                | `{} → SceneEditorStateOutput`                |
 * | `sceneeditor/opened`               | `SceneEditorOpenedParams` (notification)     |
 * | `sceneeditor/closed`               | `SceneEditorClosedParams` (notification)     |
 * | `sceneeditor/objects:changed`      | `SceneEditorObjectsChangedParams` (notify)   |
 */
export class SceneEditorPlugin {
    namespace = 'scene-editor';
    dependencies = ['input'];
    editorMeta = {
        displayName: 'Scene Editor',
        icon: 'scene-editor',
        description: 'In-engine scene editor for placing, moving, and removing scene objects.',
        events: [
            'sceneeditor/open', 'sceneeditor/close', 'sceneeditor/tool:set',
            'sceneeditor/actor-type:select',
            'sceneeditor/object:place', 'sceneeditor/object:select',
            'sceneeditor/object:move', 'sceneeditor/object:remove',
            'sceneeditor/undo', 'sceneeditor/redo',
            'sceneeditor/export', 'sceneeditor/state',
        ],
        schemas: {
            scene: {
                folder: 'scenes',
                displayName: 'Scene Layout',
            },
        },
    };
    // ── Editor state ───────────────────────────────────────────────────────────
    _core = null;
    _open = false;
    _tool = 'select';
    _selectedActorType = null;
    _selectedObjectId = null;
    // ── Objects ───────────────────────────────────────────────────────────────
    _objects = new Map();
    // ── Undo / redo stacks ────────────────────────────────────────────────────
    _undoStack = [];
    _redoStack = [];
    // ── ID counter ────────────────────────────────────────────────────────────
    _idCounter = 0;
    // ---------------------------------------------------------------------------
    // Plugin lifecycle
    // ---------------------------------------------------------------------------
    init(core) {
        this._core = core;
        core.events.on(this.namespace, 'sceneeditor/open', this._onOpen);
        core.events.on(this.namespace, 'sceneeditor/close', this._onClose);
        core.events.on(this.namespace, 'sceneeditor/tool:set', this._onToolSet);
        core.events.on(this.namespace, 'sceneeditor/actor-type:select', this._onActorTypeSelect);
        core.events.on(this.namespace, 'sceneeditor/object:place', this._onObjectPlace);
        core.events.on(this.namespace, 'sceneeditor/object:select', this._onObjectSelect);
        core.events.on(this.namespace, 'sceneeditor/object:move', this._onObjectMove);
        core.events.on(this.namespace, 'sceneeditor/object:remove', this._onObjectRemove);
        core.events.on(this.namespace, 'sceneeditor/undo', this._onUndo);
        core.events.on(this.namespace, 'sceneeditor/redo', this._onRedo);
        core.events.on(this.namespace, 'sceneeditor/export', this._onExport);
        core.events.on(this.namespace, 'sceneeditor/state', this._onState);
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
        this._open = false;
        this._tool = 'select';
        this._selectedActorType = null;
        this._selectedObjectId = null;
        this._objects.clear();
        this._undoStack = [];
        this._redoStack = [];
    }
    // ---------------------------------------------------------------------------
    // Editor control handlers
    // ---------------------------------------------------------------------------
    _onOpen = (_p) => {
        if (this._open)
            return;
        this._open = true;
        this._core?.events.emitSync('sceneeditor/opened', {});
    };
    _onClose = (_p) => {
        if (!this._open)
            return;
        this._open = false;
        this._core?.events.emitSync('sceneeditor/closed', {});
    };
    _onToolSet = (p) => {
        this._tool = p.tool;
    };
    _onActorTypeSelect = (p) => {
        this._selectedActorType = p.actorType;
    };
    _onObjectPlace = (p, output) => {
        if (!this._open)
            return;
        const newObj = {
            id: p.id ?? `obj_${this._idCounter++}`,
            actorType: this._selectedActorType ?? '',
            x: p.x,
            y: p.y,
            properties: p.properties ?? {},
        };
        this._objects.set(newObj.id, newObj);
        this._pushCommand({ edits: [{ type: 'add', prev: null, next: newObj }] });
        this._redoStack = [];
        this._core?.events.emitSync('sceneeditor/objects:changed', { objects: this._getObjectsArray() });
        output.object = newObj;
    };
    _onObjectSelect = (p) => {
        this._selectedObjectId = p.id;
    };
    _onObjectMove = (p) => {
        const existing = this._objects.get(p.id);
        if (!existing)
            return;
        const moved = { ...existing, x: p.x, y: p.y };
        this._objects.set(p.id, moved);
        this._pushCommand({ edits: [{ type: 'move', prev: existing, next: moved }] });
        this._redoStack = [];
        this._core?.events.emitSync('sceneeditor/objects:changed', { objects: this._getObjectsArray() });
    };
    _onObjectRemove = (p) => {
        const existing = this._objects.get(p.id);
        if (!existing)
            return;
        this._objects.delete(p.id);
        this._pushCommand({ edits: [{ type: 'remove', prev: existing, next: null }] });
        this._redoStack = [];
        if (this._selectedObjectId === p.id)
            this._selectedObjectId = null;
        this._core?.events.emitSync('sceneeditor/objects:changed', { objects: this._getObjectsArray() });
    };
    _onUndo = (_p) => {
        const cmd = this._undoStack.pop();
        if (!cmd)
            return;
        const reversed = [...cmd.edits].reverse();
        for (const edit of reversed) {
            if (edit.type === 'add') {
                if (edit.next)
                    this._objects.delete(edit.next.id);
            }
            else if (edit.type === 'remove') {
                if (edit.prev)
                    this._objects.set(edit.prev.id, edit.prev);
            }
            else if (edit.type === 'move') {
                if (edit.prev)
                    this._objects.set(edit.prev.id, edit.prev);
            }
        }
        this._redoStack.push(cmd);
        this._core?.events.emitSync('sceneeditor/objects:changed', { objects: this._getObjectsArray() });
    };
    _onRedo = (_p) => {
        const cmd = this._redoStack.pop();
        if (!cmd)
            return;
        for (const edit of cmd.edits) {
            if (edit.type === 'add') {
                if (edit.next)
                    this._objects.set(edit.next.id, edit.next);
            }
            else if (edit.type === 'remove') {
                if (edit.prev)
                    this._objects.delete(edit.prev.id);
            }
            else if (edit.type === 'move') {
                if (edit.next)
                    this._objects.set(edit.next.id, edit.next);
            }
        }
        this._undoStack.push(cmd);
        this._core?.events.emitSync('sceneeditor/objects:changed', { objects: this._getObjectsArray() });
    };
    _onExport = (_p, output) => {
        const objects = this._getObjectsArray();
        output.objects = objects;
        output.tiledObjectLayer = {
            type: 'objectgroup',
            name: EXPORT_LAYER_NAME,
            objects: objects.map((obj, index) => ({
                id: index + 1,
                name: obj.id,
                type: obj.actorType,
                x: obj.x,
                y: obj.y,
                properties: Object.entries(obj.properties).map(([name, value]) => ({
                    name,
                    type: 'string',
                    value: String(value),
                })),
            })),
        };
    };
    _onState = (_p, output) => {
        output.open = this._open;
        output.tool = this._tool;
        output.selectedActorType = this._selectedActorType;
        output.selectedObjectId = this._selectedObjectId;
        output.canUndo = this._undoStack.length > 0;
        output.canRedo = this._redoStack.length > 0;
    };
    // ---------------------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------------------
    _getObjectsArray() {
        return Array.from(this._objects.values());
    }
    _pushCommand(cmd) {
        this._undoStack.push(cmd);
        if (this._undoStack.length > MAX_UNDO_HISTORY) {
            this._undoStack.shift();
        }
    }
}
