// @inkshot/engine/dist/plugins/physics/RapierPhysicsAdapter.js
import { CollisionLayer } from '../../types/physics.js';
// ---------------------------------------------------------------------------
// Geometry helpers
// ---------------------------------------------------------------------------
function getShapeAABB(shape, px, py) {
    const ox = shape.offsetX ?? 0;
    const oy = shape.offsetY ?? 0;
    if (shape.type === 'rect') {
        return { left: px + ox, top: py + oy, right: px + ox + shape.width, bottom: py + oy + shape.height };
    }
    if (shape.type === 'circle') {
        return { left: px + ox - shape.radius, top: py + oy - shape.radius, right: px + ox + shape.radius, bottom: py + oy + shape.radius };
    }
    return { left: px + ox, top: py + oy, right: px + ox, bottom: py + oy };
}
function aabbsOverlap(a, b) {
    return a.left < b.right && a.right > b.left && a.top < b.bottom && a.bottom > b.top;
}
function rayVsAABB(ox, oy, dx, dy, left, top, right, bottom) {
    let tMin = 0;
    let tMax = Infinity;
    if (Math.abs(dx) > 1e-9) {
        const t1 = (left - ox) / dx;
        const t2 = (right - ox) / dx;
        tMin = Math.max(tMin, Math.min(t1, t2));
        tMax = Math.min(tMax, Math.max(t1, t2));
    }
    else if (ox < left || ox > right) {
        return null;
    }
    if (Math.abs(dy) > 1e-9) {
        const t1 = (top - oy) / dy;
        const t2 = (bottom - oy) / dy;
        tMin = Math.max(tMin, Math.min(t1, t2));
        tMax = Math.min(tMax, Math.max(t1, t2));
    }
    else if (oy < top || oy > bottom) {
        return null;
    }
    return tMin <= tMax ? tMin : null;
}
// ---------------------------------------------------------------------------
// RapierPhysicsAdapter
// ---------------------------------------------------------------------------
/**
 * **Rapier.js** WASM physics backend for Inkshot Engine.
 *
 * Wraps the [Rapier.js](https://rapier.rs/) high-performance rigid-body
 * simulation library (2-D) and exposes it through the unified `physics/*`
 * EventBus API so it can replace {@link KinematicPhysicsAdapter} without any
 * changes to game code.
 *
 * Rapier must be installed separately and the WASM module must be **fully
 * initialised** before constructing this adapter:
 *
 * ```sh
 * npm install @dimforge/rapier2d
 * ```
 *
 * ```ts
 * import RAPIER from '@dimforge/rapier2d';
 * import { createEngine, RapierPhysicsAdapter } from 'inkshot-engine';
 *
 * await RAPIER.init();
 * const { core } = await createEngine({
 *   plugins: [new RapierPhysicsAdapter(RAPIER)],
 * });
 * ```
 *
 * ### Body modes
 * - Bodies registered with {@link CollisionLayer.BODY} use **kinematic
 *   position-based** rigid bodies, controlled via `physics/move`.
 * - Non-BODY layers (HITBOX, HURTBOX, SENSOR) use **dynamic** rigid bodies
 *   so they respond to `physics/impulse`.
 *
 * ### Tilemap static bodies
 * `physics/tilemap:set` creates a batch of static (`fixed`) rectangular
 * colliders for every non-empty tile, replacing the previous set.
 *
 * ### Entity position sync
 * After each `core/update` tick the adapter reads every dynamic body's
 * translation and writes it back to `entity.position`, keeping the logical
 * and visual state in lock-step.
 *
 * @public
 */
export class RapierPhysicsAdapter {
    namespace = 'physics';
    /** Must be initialised after EntityManager so `entity/query` is available. */
    dependencies = ['entityManager'];
    editorMeta = {
        displayName: 'Rapier Physics Adapter',
        icon: 'physics',
        description: 'Physics backend powered by Rapier (WASM) for high-performance simulation.',
        events: [
            'physics/body:add', 'physics/body:remove', 'physics/move', 'physics/impulse',
            'physics/query', 'physics/raycast', 'physics/hit', 'physics/overlap',
            'physics/tilemap:set', 'physics/grid:snap',
        ],
    };
    _core = null;
    _RAPIER;
    _options;
    _world;
    /** entityId → { rapier body, rapier collider, shape, layer }. */
    _records = new Map();
    /** Static bodies created from the current tilemap. */
    _tilemapBodies = [];
    /** Active tilemap data (for grid utilities). */
    _tilemap = null;
    /** Solid-tile lookup. Key: `"row,col"`. */
    _solidTiles = new Set();
    /** Handle → entityId reverse lookup for query/raycast results. */
    _colliderHandleToEntityId = new Map();
    /** Hit pairs active in the previous update. */
    _prevHitPairs = new Set();
    /** Sensor overlap pairs active in the previous update. */
    _prevSensorPairs = new Set();
    constructor(RAPIER, options = {}) {
        this._RAPIER = RAPIER;
        this._options = {
            gravity: options.gravity ?? { x: 0, y: 9.81 },
        };
    }
    // ---------------------------------------------------------------------------
    // EnginePlugin lifecycle
    // ---------------------------------------------------------------------------
    init(core) {
        this._core = core;
        const { events } = core;
        this._world = new this._RAPIER.World(this._options.gravity);
        events.on(this.namespace, 'physics/body:add', (params) => {
            this._addBody(params);
        });
        events.on(this.namespace, 'physics/body:remove', (params) => {
            this._removeBody(params.entityId);
        });
        events.on(this.namespace, 'physics/tilemap:set', (params) => {
            this._setTilemap(params);
        });
        events.on(this.namespace, 'physics/move', (params, output) => {
            const result = this._move(params.entityId, params.dx, params.dy);
            output.x = result.x;
            output.y = result.y;
            output.blockedX = result.blockedX;
            output.blockedY = result.blockedY;
        });
        events.on(this.namespace, 'physics/impulse', (params) => {
            this._applyImpulse(params.entityId, params.forceX, params.forceY);
        });
        events.on(this.namespace, 'physics/query', (params, output) => {
            output.entities = this._query(params);
        });
        events.on(this.namespace, 'physics/raycast', (params, output) => {
            const result = this._raycast(params);
            output.hit = result.hit;
            if (result.entityId !== undefined)
                output.entityId = result.entityId;
            if (result.point !== undefined)
                output.point = result.point;
            if (result.distance !== undefined)
                output.distance = result.distance;
            if (result.tileHit !== undefined)
                output.tileHit = result.tileHit;
        });
        events.on(this.namespace, 'physics/grid:snap', (params, output) => {
            const s = this._gridSnap(params.x, params.y);
            output.x = s.x;
            output.y = s.y;
        });
        events.on(this.namespace, 'physics/grid:worldToTile', (params, output) => {
            const t = this._worldToTile(params.x, params.y);
            output.col = t.col;
            output.row = t.row;
        });
        events.on(this.namespace, 'physics/grid:tileToWorld', (params, output) => {
            const w = this._tileToWorld(params.col, params.row);
            output.x = w.x;
            output.y = w.y;
        });
        // Step the physics world and sync entity positions each fixed update.
        events.on(this.namespace, 'core/update', () => {
            this._world.step();
            this._syncPositions();
            this._detectOverlaps();
        }, { priority: -20 });
        // Auto-remove bodies when their owning entities are destroyed.
        events.on(this.namespace, 'entity/destroyed', ({ entity }) => {
            this._removeBody(entity.id);
        });
    }
    destroy(core) {
        core.events.removeNamespace(this.namespace);
        this._records.clear();
        this._tilemapBodies = [];
        this._solidTiles.clear();
        this._colliderHandleToEntityId.clear();
        this._tilemap = null;
        this._prevHitPairs.clear();
        this._prevSensorPairs.clear();
        this._core = null;
    }
    // ---------------------------------------------------------------------------
    // Private: body management
    // ---------------------------------------------------------------------------
    _addBody(params) {
        const entityMap = this._getAllEntities();
        const entity = entityMap.get(params.entityId);
        const pos = entity?.position ?? { x: 0, y: 0 };
        const shape = params.shape;
        const isKinematic = !!(params.layer & CollisionLayer.BODY);
        const bodyDesc = isKinematic
            ? this._RAPIER.RigidBodyDesc.kinematicPositionBased()
            : this._RAPIER.RigidBodyDesc.dynamic();
        bodyDesc.setTranslation(pos.x + (shape.offsetX ?? 0) + (shape.type === 'rect' ? shape.width / 2 : 0), pos.y + (shape.offsetY ?? 0) + (shape.type === 'rect' ? shape.height / 2 : 0));
        const body = this._world.createRigidBody(bodyDesc);
        const colliderDesc = this._buildColliderDesc(shape);
        const collider = this._world.createCollider(colliderDesc, body);
        this._records.set(params.entityId, { body, collider, shape, layer: params.layer });
        this._colliderHandleToEntityId.set(collider.handle, params.entityId);
    }
    _removeBody(entityId) {
        const record = this._records.get(entityId);
        if (record) {
            this._colliderHandleToEntityId.delete(record.collider.handle);
            this._world.removeCollider(record.collider, false);
            this._world.removeRigidBody(record.body);
            this._records.delete(entityId);
        }
        for (const key of this._prevHitPairs) {
            if (key.startsWith(entityId + '|') || key.endsWith('|' + entityId)) {
                this._prevHitPairs.delete(key);
            }
        }
        for (const key of this._prevSensorPairs) {
            if (key.startsWith(entityId + '|') || key.endsWith('|' + entityId)) {
                this._prevSensorPairs.delete(key);
            }
        }
    }
    _buildColliderDesc(shape) {
        if (shape.type === 'rect') {
            return this._RAPIER.ColliderDesc.cuboid(shape.width / 2, shape.height / 2);
        }
        if (shape.type === 'circle') {
            return this._RAPIER.ColliderDesc.ball(shape.radius);
        }
        // point — represent as a tiny ball
        return this._RAPIER.ColliderDesc.ball(0.5);
    }
    // ---------------------------------------------------------------------------
    // Private: tilemap
    // ---------------------------------------------------------------------------
    _setTilemap(params) {
        // Remove old tilemap bodies.
        for (const b of this._tilemapBodies) {
            this._world.removeRigidBody(b);
        }
        this._tilemapBodies = [];
        this._solidTiles.clear();
        this._tilemap = {
            tileSize: params.tileSize,
            layers: params.layers,
            tileShapes: params.tileShapes,
        };
        const ts = params.tileSize;
        const newBodies = [];
        for (let row = 0; row < params.layers.length; row++) {
            const rowData = params.layers[row];
            for (let col = 0; col < rowData.length; col++) {
                const tileValue = rowData[col];
                const tileShape = params.tileShapes[tileValue];
                if (!tileShape || tileShape === 'empty')
                    continue;
                this._solidTiles.add(`${row},${col}`);
                const cx = col * ts + ts / 2;
                const cy = row * ts + ts / 2;
                const bodyDesc = this._RAPIER.RigidBodyDesc.fixed().setTranslation(cx, cy);
                const body = this._world.createRigidBody(bodyDesc);
                const colliderDesc = this._RAPIER.ColliderDesc.cuboid(ts / 2, ts / 2);
                this._world.createCollider(colliderDesc, body);
                newBodies.push(body);
            }
        }
        this._tilemapBodies = newBodies;
    }
    // ---------------------------------------------------------------------------
    // Private: movement
    // ---------------------------------------------------------------------------
    _move(entityId, dx, dy) {
        const record = this._records.get(entityId);
        const entityMap = this._getAllEntities();
        const entity = entityMap.get(entityId);
        if (!record || !(record.layer & CollisionLayer.BODY) || !entity) {
            return {
                x: entity?.position.x ?? 0,
                y: entity?.position.y ?? 0,
                blockedX: false,
                blockedY: false,
            };
        }
        let posX = entity.position.x + dx;
        let posY = entity.position.y + dy;
        let blockedX = false;
        let blockedY = false;
        if (this._tilemap && dx !== 0) {
            const res = this._checkTileBlockingX(record.shape, posX, posY, dx);
            if (res.blocked) {
                posX = res.resolved;
                blockedX = true;
            }
        }
        if (this._tilemap && dy !== 0) {
            const res = this._checkTileBlockingY(record.shape, posX, posY, dy);
            if (res.blocked) {
                posY = res.resolved;
                blockedY = true;
            }
        }
        entity.position.x = posX;
        entity.position.y = posY;
        // Move the Rapier kinematic body.
        const ox = record.shape.offsetX ?? 0;
        const oy = record.shape.offsetY ?? 0;
        let bodyX = posX + ox;
        let bodyY = posY + oy;
        if (record.shape.type === 'rect') {
            bodyX += record.shape.width / 2;
            bodyY += record.shape.height / 2;
        }
        record.body.setTranslation({ x: bodyX, y: bodyY }, true);
        return { x: posX, y: posY, blockedX, blockedY };
    }
    _checkTileBlockingX(shape, posX, posY, dx) {
        if (!this._tilemap)
            return { blocked: false, resolved: posX };
        const ts = this._tilemap.tileSize;
        const aabb = getShapeAABB(shape, posX, posY);
        const rowMin = Math.floor(aabb.top / ts);
        const rowMax = Math.floor((aabb.bottom - 1) / ts);
        const colMin = Math.floor(aabb.left / ts);
        const colMax = Math.floor((aabb.right - 1) / ts);
        let blocked = false;
        let resolved = posX;
        const ox = shape.offsetX ?? 0;
        const rightOff = shape.type === 'rect' ? ox + shape.width : shape.type === 'circle' ? ox + shape.radius : ox;
        const leftOff = shape.type === 'circle' ? ox - shape.radius : ox;
        for (let row = rowMin; row <= rowMax; row++) {
            for (let col = colMin; col <= colMax; col++) {
                if (!this._solidTiles.has(`${row},${col}`))
                    continue;
                blocked = true;
                if (dx > 0) {
                    const snap = col * ts - rightOff;
                    if (snap < resolved)
                        resolved = snap;
                }
                else {
                    const snap = (col + 1) * ts - leftOff;
                    if (snap > resolved)
                        resolved = snap;
                }
            }
        }
        return { blocked, resolved };
    }
    _checkTileBlockingY(shape, posX, posY, dy) {
        if (!this._tilemap)
            return { blocked: false, resolved: posY };
        const ts = this._tilemap.tileSize;
        const aabb = getShapeAABB(shape, posX, posY);
        const rowMin = Math.floor(aabb.top / ts);
        const rowMax = Math.floor((aabb.bottom - 1) / ts);
        const colMin = Math.floor(aabb.left / ts);
        const colMax = Math.floor((aabb.right - 1) / ts);
        let blocked = false;
        let resolved = posY;
        const oy = shape.offsetY ?? 0;
        const botOff = shape.type === 'rect' ? oy + shape.height : shape.type === 'circle' ? oy + shape.radius : oy;
        const topOff = shape.type === 'circle' ? oy - shape.radius : oy;
        for (let row = rowMin; row <= rowMax; row++) {
            for (let col = colMin; col <= colMax; col++) {
                if (!this._solidTiles.has(`${row},${col}`))
                    continue;
                blocked = true;
                if (dy > 0) {
                    const snap = row * ts - botOff;
                    if (snap < resolved)
                        resolved = snap;
                }
                else {
                    const snap = (row + 1) * ts - topOff;
                    if (snap > resolved)
                        resolved = snap;
                }
            }
        }
        return { blocked, resolved };
    }
    // ---------------------------------------------------------------------------
    // Private: impulse
    // ---------------------------------------------------------------------------
    _applyImpulse(entityId, forceX, forceY) {
        const record = this._records.get(entityId);
        if (!record)
            return;
        record.body.applyImpulse({ x: forceX, y: forceY }, true);
    }
    // ---------------------------------------------------------------------------
    // Private: spatial query (AABB overlap using internal records)
    // ---------------------------------------------------------------------------
    _query(params) {
        const results = [];
        const queryAABB = getShapeAABB(params.shape, params.position.x, params.position.y);
        const entityMap = this._getAllEntities();
        // Use the Rapier world intersectionsWithShape if available, otherwise fall
        // back to a manual AABB sweep over registered bodies.
        if (typeof this._world.intersectionsWithShape === 'function') {
            const shape = this._buildQueryShape(params.shape);
            this._world.intersectionsWithShape(params.position, 0, shape, (handle) => {
                const entityId = this._colliderHandleToEntityId.get(handle);
                if (!entityId)
                    return true;
                if (params.excludeEntityId === entityId)
                    return true;
                const record = this._records.get(entityId);
                if (!record)
                    return true;
                if (!(record.layer & params.layerMask))
                    return true;
                results.push(entityId);
                return true; // continue iteration
            });
        }
        else {
            // Fallback: manual AABB check.
            for (const [entityId, record] of this._records) {
                if (params.excludeEntityId === entityId)
                    continue;
                if (!(record.layer & params.layerMask))
                    continue;
                const entity = entityMap.get(entityId);
                if (!entity)
                    continue;
                const entityAABB = getShapeAABB(record.shape, entity.position.x, entity.position.y);
                if (aabbsOverlap(queryAABB, entityAABB)) {
                    results.push(entityId);
                }
            }
        }
        return results;
    }
    // ---------------------------------------------------------------------------
    // Private: raycast
    // ---------------------------------------------------------------------------
    _raycast(params) {
        const { origin, direction, maxDistance = 10_000, layerMask } = params;
        const len = Math.sqrt(direction.x * direction.x + direction.y * direction.y);
        if (len < 1e-9)
            return { hit: false };
        const ndx = direction.x / len;
        const ndy = direction.y / len;
        let bestDist = maxDistance;
        let bestEntityId;
        let tileHit = false;
        if (typeof this._world.castShape === 'function') {
            // Use a tiny ball as the sweep shape for the ray query.
            const sweepShape = this._RAPIER.ColliderDesc.ball(0.5);
            const hit = this._world.castShape(origin, 0, { x: ndx, y: ndy }, sweepShape, maxDistance, false);
            if (hit && hit.toi < bestDist) {
                bestDist = hit.toi;
                if (hit.colliderHandle !== undefined) {
                    const eid = this._colliderHandleToEntityId.get(hit.colliderHandle);
                    if (eid) {
                        bestEntityId = eid;
                    }
                    else {
                        tileHit = true;
                    }
                }
            }
        }
        else {
            // Fallback: manual AABB ray check.
            const entityMap = this._getAllEntities();
            for (const [entityId, record] of this._records) {
                if (!(record.layer & layerMask))
                    continue;
                const entity = entityMap.get(entityId);
                if (!entity)
                    continue;
                const aabb = getShapeAABB(record.shape, entity.position.x, entity.position.y);
                const t = rayVsAABB(origin.x, origin.y, ndx, ndy, aabb.left, aabb.top, aabb.right, aabb.bottom);
                if (t !== null && t < bestDist) {
                    bestDist = t;
                    bestEntityId = entityId;
                    tileHit = false;
                }
            }
            // Check solid tiles.
            if (this._tilemap) {
                const ts = this._tilemap.tileSize;
                for (const key of this._solidTiles) {
                    const sep = key.indexOf(',');
                    const row = parseInt(key.slice(0, sep), 10);
                    const col = parseInt(key.slice(sep + 1), 10);
                    const tl = col * ts;
                    const tt = row * ts;
                    const t = rayVsAABB(origin.x, origin.y, ndx, ndy, tl, tt, tl + ts, tt + ts);
                    if (t !== null && t < bestDist) {
                        bestDist = t;
                        bestEntityId = undefined;
                        tileHit = true;
                    }
                }
            }
        }
        if (!bestEntityId && !tileHit)
            return { hit: false };
        return {
            hit: true,
            entityId: bestEntityId,
            point: { x: origin.x + ndx * bestDist, y: origin.y + ndy * bestDist },
            distance: bestDist,
            tileHit,
        };
    }
    /** Build a Rapier shape descriptor for a given ColliderShape (used in queries). */
    _buildQueryShape(shape) {
        if (shape.type === 'rect') {
            return this._RAPIER.ColliderDesc.cuboid(shape.width / 2, shape.height / 2);
        }
        if (shape.type === 'circle') {
            return this._RAPIER.ColliderDesc.ball(shape.radius);
        }
        return this._RAPIER.ColliderDesc.ball(0.5);
    }
    // ---------------------------------------------------------------------------
    // Private: grid utilities
    // ---------------------------------------------------------------------------
    _gridSnap(x, y) {
        const ts = this._tilemap?.tileSize ?? 0;
        if (!ts)
            return { x, y };
        return { x: Math.round(x / ts) * ts, y: Math.round(y / ts) * ts };
    }
    _worldToTile(x, y) {
        const ts = this._tilemap?.tileSize ?? 1;
        return { col: Math.floor(x / ts), row: Math.floor(y / ts) };
    }
    _tileToWorld(col, row) {
        const ts = this._tilemap?.tileSize ?? 0;
        return { x: col * ts, y: row * ts };
    }
    // ---------------------------------------------------------------------------
    // Private: position sync and overlap detection
    // ---------------------------------------------------------------------------
    /**
     * After each Rapier world step, sync body translations back to entity.position.
     */
    _syncPositions() {
        if (!this._core)
            return;
        const entityMap = this._getAllEntities();
        for (const [entityId, record] of this._records) {
            // Only sync dynamic bodies — kinematic BODY bodies are moved via physics/move.
            if (record.layer & CollisionLayer.BODY)
                continue;
            const entity = entityMap.get(entityId);
            if (!entity)
                continue;
            const t = record.body.translation();
            const ox = record.shape.offsetX ?? 0;
            const oy = record.shape.offsetY ?? 0;
            let entityX = t.x - ox;
            let entityY = t.y - oy;
            if (record.shape.type === 'rect') {
                entityX -= record.shape.width / 2;
                entityY -= record.shape.height / 2;
            }
            entity.position.x = entityX;
            entity.position.y = entityY;
        }
    }
    _detectOverlaps() {
        if (!this._core || this._records.size === 0)
            return;
        const entityMap = this._getAllEntities();
        const hitboxers = [];
        const hurtboxers = [];
        const sensorers = [];
        for (const [id, record] of this._records) {
            const entity = entityMap.get(id);
            if (!entity)
                continue;
            const pos = entity.position;
            if (record.layer & CollisionLayer.HITBOX)
                hitboxers.push({ id, layer: record.layer, shape: record.shape, pos });
            if (record.layer & CollisionLayer.HURTBOX)
                hurtboxers.push({ id, layer: record.layer, shape: record.shape, pos });
            if (record.layer & CollisionLayer.SENSOR)
                sensorers.push({ id, layer: record.layer, shape: record.shape, pos });
        }
        const currentHitPairs = new Set();
        for (const atk of hitboxers) {
            for (const vic of hurtboxers) {
                if (atk.id === vic.id)
                    continue;
                const atkAABB = getShapeAABB(atk.shape, atk.pos.x, atk.pos.y);
                const vicAABB = getShapeAABB(vic.shape, vic.pos.x, vic.pos.y);
                if (!aabbsOverlap(atkAABB, vicAABB))
                    continue;
                const key = `${atk.id}|${vic.id}`;
                currentHitPairs.add(key);
                if (!this._prevHitPairs.has(key)) {
                    this._core.events.emitSync('physics/hit', { attackerId: atk.id, victimId: vic.id });
                }
            }
        }
        this._prevHitPairs.clear();
        for (const key of currentHitPairs)
            this._prevHitPairs.add(key);
        const currentSensorPairs = new Set();
        for (let i = 0; i < sensorers.length; i++) {
            for (let j = i + 1; j < sensorers.length; j++) {
                const a = sensorers[i];
                const b = sensorers[j];
                const aAABB = getShapeAABB(a.shape, a.pos.x, a.pos.y);
                const bAABB = getShapeAABB(b.shape, b.pos.x, b.pos.y);
                if (!aabbsOverlap(aAABB, bAABB))
                    continue;
                const key = a.id < b.id ? `${a.id}|${b.id}` : `${b.id}|${a.id}`;
                currentSensorPairs.add(key);
            }
        }
        for (const key of currentSensorPairs) {
            if (!this._prevSensorPairs.has(key)) {
                const sep = key.indexOf('|');
                this._core.events.emitSync('physics/overlap', { entityAId: key.slice(0, sep), entityBId: key.slice(sep + 1), entered: true });
            }
        }
        for (const key of this._prevSensorPairs) {
            if (!currentSensorPairs.has(key)) {
                const sep = key.indexOf('|');
                this._core.events.emitSync('physics/overlap', { entityAId: key.slice(0, sep), entityBId: key.slice(sep + 1), entered: false });
            }
        }
        this._prevSensorPairs.clear();
        for (const key of currentSensorPairs)
            this._prevSensorPairs.add(key);
    }
    // ---------------------------------------------------------------------------
    // Private: entity helpers
    // ---------------------------------------------------------------------------
    _getAllEntities() {
        if (!this._core)
            return new Map();
        const { output } = this._core.events.emitSync('entity/query', {});
        const map = new Map();
        for (const e of output.entities ?? []) {
            map.set(e.id, e);
        }
        return map;
    }
}
